---
title: "OSHetero2021"
author: "Sanjana Rajan, Emily Franz, Matthew Cannon"
date: "`r format(Sys.time(), '%m/%d/%Y')`"
output: 
  html_document:
    toc: true
    toc_float: true
    toc_depth: 5
    number_sections: false
    code_folding: hide
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
  tidy = TRUE,
  echo = TRUE,
  cache = TRUE,
  collapse = TRUE,
  tidy.opts = list(width.cutoff = 95),
  message = FALSE,
  warning = FALSE,
  cache.lazy = FALSE,
  fig.show='hide'
)
```

*Seurat object loading and processing code can be found in "OSHetero2021_Load_Process.Rmd". Object outputs from that code are loaded in the code below.*

# Figure 1

**Cell line and PDX models of osteosarcoma display transcriptional heterogeneity.**

```{r Fig1_lib, cache=FALSE}
source("Downstream.v2.R")
set.seed(888)

# loading libraries
library(Seurat)
library(ggplot2)
library(msigdbr)
library(clusterProfiler)
library(grid)
library(data.table)
library(tidyverse)
library(ggridges)

message(paste("Working directory: ", getwd()))
```

## Figure 1A

A) UMAP analysis of OS-17 cells (n = 3178) grown in cell culture. Cell cycle distribution of cells (n = 3178) is visualized as pie charts.

```{r Fig1a_OS17}
load("Data/os17_cx_CCR.RData")

# Culture plot
figure1_a1 <- DimPlot(os17_cx,
    reduction = "umap",
    pt.size = 1,
    label = T) +
  coord_fixed() +
  ggtitle("OS17 in Culture") +
  NoAxes() +
  scale_color_npg(alpha = 1)

# View Culture plot in Rmd
figure1_a1

# Save as pdf
if(!dir.exists("Figures")) {
  dir.create("Figures")
}
if(!dir.exists("Figures/figure_1")) {
  dir.create("Figures/figure_1")
}
pdf("Figures/figure_1/figure_1a_os17.pdf",
    width = 5,
    height = 5)
figure1_a1
dev.off()


# Cell cycle distribution of clusters
cid <- sort(unique(os17_cx@active.ident))

vplayout <- function(x, y) viewport(layout.pos.row = x,
                                    layout.pos.col = y)

pdf("Figures/figure_1/figure_1a_os17_pie.pdf",
    width = 6,
    height = 2)

grid.newpage()
pushViewport(viewport(layout = grid.layout(1, length(unique(os17_cx@active.ident)))))

for(i in seq_along(unique(os17_cx@active.ident))){
  temp <- WhichCells(os17_cx,
                     idents = cid[[i]])
  LT.cells <- subset(os17_cx,
                     cells = temp)
  df <- table(LT.cells$Phase)
  bp <- ggplot(as.data.frame(df),
               aes(x = "",
                   y = Freq,
                   fill = Var1)) +
    geom_bar(width = 1,
             stat = "identity")
  pie <- bp +
    coord_polar("y",
                start = 0) +
    scale_fill_brewer(palette = "Blues") +
    theme_minimal() +
    NoAxes() + NoLegend() +
    ggtitle(paste("Cluster",
                  cid[[i]]))
  print(pie,
        vp = vplayout(ceiling(i / length(unique(os17_cx@active.ident))),
                      i))
}
dev.off()
```

## Figure 1B

B) UMAP analysis of NCH-OS-7 cells (n = 1998) grown as subcutaneous flank tumor. Cell cycle distribution of cells is visualized as pie charts.

```{r Fig1b_NCHOS7}
load(file = "Data/os7_cx_CCR.RData")

# Flank plots

figure_1b <- DimPlot(os7_cx,
                     reduction = "umap",
                     pt.size = 1,
                     label = T) +
  coord_fixed() +
  ggtitle("NCHOS7 in Flank") +
  NoAxes() +
  scale_color_npg(alpha = 1)

# View in Rmd
figure_1b

# Save as pdf
pdf("Figures/figure_1/figure_1b_nchos7.pdf",
    width = 5,
    height = 5)
figure_1b
dev.off()

# Cell cycle distribution of clusters
cid <- sort(unique(os7_cx@active.ident))

vplayout <- function(x, y) viewport(layout.pos.row = x, layout.pos.col = y)

pdf("Figures/figure_1/figure_1b_nchos7_pie.pdf",
    width = 6,
    height = 2)

grid.newpage()
pushViewport(viewport(layout = grid.layout(1, length(unique(os7_cx@active.ident)))))
for(i in seq_along(unique(os7_cx@active.ident))){
  temp <- WhichCells(os7_cx,
                     idents = cid[[i]]) 
  LT.cells <- subset(os7_cx,
                     cells = temp)
  df <- table(LT.cells$Phase)
  bp <- ggplot(as.data.frame(df),
               aes(x="",
                   y=Freq,
                   fill=Var1)) +
    geom_bar(width = 1,
             stat = "identity")
  pie <- bp +
    coord_polar("y",
                start=0) +
    scale_fill_brewer(palette="Blues") +
    theme_minimal() +
    NoAxes() +
    NoLegend() +
    ggtitle(paste("Cluster",
                  cid[[i]]))
  
  print(pie,
        vp=vplayout(ceiling(i/length(unique(os7_cx@active.ident))),
                    i))
}
dev.off()
```

## Figure 1C

C) Pathway enrichment analysis for hallmark gene sets associated with genes upregulated in distinct clusters identified in the two osteosarcoma models. P values were adjusted for multiple comparisons. Boxes in grey identify non-significant enrichments, whereas boxes in red identify statistically significant enrichments. MTORC1: mechanistic target of rapamycin (mTOR) complex 1. PI3K: Phosphoinositide 3-kinase. AKT: Protein kinase B. TGF: Transforming growth-factor. TNF: tumor necrosis factor. NFKB: Nuclear Factor kappa-light-chain-enhancer of activated B cells.

```{r Fig1c_dgea, dependson='Fig1b_NCHOS7'}
# Pathway enrichment analysis
# Display adjust p-values
B.list <- list(OS17 = os17_cx,
               NCHOS7 = os7_cx)

em.hm.list <- list()

for (i in seq_along(B.list)) {
  em.hm.list[[i]] <- DGEA(B.list[[i]])
}

for(i in seq_along(em.hm.list)) {
  em.hm.list[[i]] <- setDT(em.hm.list[[i]],
                           keep.rownames = TRUE)[]
}

temp1 <- em.hm.list[[1]]
temp2 <- em.hm.list[[2]]
cx.pd <- full_join(temp1,
                   temp2,
                   by = "rn")

cx.pd <- column_to_rownames(cx.pd,
                            var = "rn")
cx.pd[is.na(cx.pd)] <- 1

#remove "HALLMARK_"
c <- rownames(cx.pd)
c <- gsub("HALLMARK_",
          "",
          c)

#remove "_" by removing special characters
c <- gsub("_",
          " ",
          c)
rownames(cx.pd) <- c

# Take the -log10
cx.pd.log <- -log10(cx.pd)
colnames(cx.pd.log) <- c("A0",
                         "A1",
                         "A2",
                         "A3",
                         "B0",
                         "B1",
                         "B2",
                         "B3")

# To prevent values that show up as -0.00, we take the absolute values
cx.pd.log <- abs(cx.pd.log)

figure_1c <- cx.pd.log %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column(var = "Sample") %>%
  pivot_longer(c(-Sample),
               names_to = "Pathway",
               values_to = "pval",
               names_repair = "minimal") %>%
  mutate(Signif = pval >= (-1 * log10(0.05))) %>%
  ggplot(.,
         aes(x = Sample,
             y = Pathway,
             fill = Signif)) +
  geom_tile() +
  geom_text(aes(label = sprintf("%0.2f",
                                pval))) + 
  scale_fill_manual(values = c("gray",
                               "red")) +
  scale_y_discrete(limits = rev,
                   position = "right") +
  theme(legend.position = "right",
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        axis.ticks.x = element_blank()) +
  ylab("")
figure_1c

#Save as PDF
pdf("Figures/figure_1/figure_1c.pdf",
    width = 8,
    height = 8)
figure_1c
dev.off()
```

# Figure 2

Figure 2. Osteosarcoma cells adopt distinct transcriptional profiles as they colonize tibia and lung microenvironments. A) Schematic of study design depicting generation of orthotopic primary and metastatic tumors. Tumors were harvested for scRNA-seq when mice reached endpoint.

## Figure 2B

B) Venn diagrams showing overlap of differentially expressed genes that are up- or down- regulated in primary or metastatic tumors relative to corresponding starting population of cells (cell line or PDX flank tumors). Region outlined in red identifies differentially expressed genes shared across at least three models.

```{r Fig2b, fig.height=10, fig.width=10}
load("Data/OS_merged_postCCR.RData")
# Set seed before subsetting and DE analysis to ensure consistent results
set.seed(100)

OS$tissue <- OS$src %>%
  stringr::str_replace("Flank", "Culture")

OS <- subset(OS, subset = model != "Osteoblasts")

Idents(OS) <- OS$tissue

de_out <- list()
for (tissue_name in c("Tibia", "Lung")) {
  for (cl_name in unique(OS$model)) {
    message(tissue_name, " ", cl_name)
    temp_seurat <- subset(OS, subset = (tissue == tissue_name |
                                          tissue == "Culture") &
                            model == cl_name)
    temp_de_out <-
      FindMarkers(temp_seurat,
                  ident.1 = tissue_name,
                  ident.2 = "Culture",
                  min.pct = 0.1) %>%
      filter(p_val_adj <= 0.05)
    
    de_out[[tissue_name]][["up"]][[cl_name]] <-
      temp_de_out %>%
      filter(avg_log2FC > 0) %>%
      rownames()
    
    de_out[[tissue_name]][["down"]][[cl_name]] <-
      temp_de_out %>%
      filter(avg_log2FC < 0) %>%
      rownames()
    
    rm(temp_seurat, temp_de_out)
  }
}

figure_2b <- list()
for (tissue_name in c("Tibia", "Lung")) {
  for (direction in c("up", "down")) {
    figure_2b[[paste0(tissue_name,
                      " - ",
                      direction)]] <-
      ggvenn::ggvenn(de_out[[tissue_name]][[direction]],
                     fill_color = c("#0073C2FF",
                                    "#EFC000FF",
                                    "#868686FF",
                                    "#CD534CFF"),
                     stroke_size = 0.1,
                     set_name_size = 2,
                     text_size = 1.3,
                     show_percentage = TRUE) +
      ggtitle(paste0(tissue_name,
                     " - ",
                     direction)) +
      theme(plot.title = element_text(hjust = 0.5,
                                      vjust = -8,
                                      size = 10,
                                      face = "bold"),
            plot.margin = margin(-2, 0, -2, 0, "cm"))
    
    # Fix the position of the bottom two set names
    figure_2b[[paste0(tissue_name,
                      " - ",
                      direction)]]$layers[[3]]$data$x <- 
      c(-1, -0.8, 0.8, 1)
  }
}

#Save as PDF
if(!dir.exists("Figures/figure_2")) {
  dir.create("Figures/figure_2")
}
pdf("Figures/figure_2/figure_2b.pdf",
    width = 4,
    height = 4)
# Put all four plots into one image
gridExtra::grid.arrange(grobs = figure_2b,
                        vp = viewport(width = 1.0, height = 0.9))
dev.off()

# Save RData obj
save(figure_2b, file = "Figures/figure_2/figure_2b.RData")
```

## Figure 2C

C) Pathway enrichment analysis with adjusted p values for hallmark gene sets associated with these shared genes

```{r Fig2c, dependson = "Fig2b", fig.height=6, fig.width=10}
### Extract genes that are shared in these datasets
load("Data/OS_merged_postCCR.RData")
# Set seed before subsetting and DE analysis to ensure consistent results
set.seed(100)

data <- list(
  os17 = subset(OS, cells = WhichCells(OS, expression = model == "OS-17")),
  t143B = subset(OS, cells = WhichCells(OS, expression = model == "143B")),
  NCHOS2 = subset(OS, cells = WhichCells(OS, expression = model == "NCH-OS-2")),
  NCHOS7 = subset(OS, cells = WhichCells(OS, expression = model == "NCH-OS-7"))
)

# Rename "Flank" to "Culture" for NCHOS samples
data$NCHOS2$src[grep("Flank", data$NCHOS2$src)] <- "Culture"
data$NCHOS7$src[grep("Flank", data$NCHOS7$src)] <- "Culture"

fig_2c_data <- tibble()

for (tissue_name in c("Tibia", "Lung")) {
  for (direction in c("up", "down")) {
    de_genes <- list()
    for (i in seq_len(length(data))) {
      Idents(data[[i]]) <- data[[i]]$src
      de_genes[[i]] <-
        FindMarkers(data[[i]],
                    ident.1 = tissue_name,
                    ident.2 = "Culture",
                    only.pos = FALSE,
                    min.pct = 0.1)
      
      # Select genes that change in the direction of interest
      if (direction == "up") {
        de_genes[[i]] <-
          de_genes[[i]][de_genes[[i]]$p_val_adj < 0.05 &
                          de_genes[[i]]$avg_log2FC > 0, ]
      } else {
        de_genes[[i]] <-
          de_genes[[i]][de_genes[[i]]$p_val_adj < 0.05 &
                          de_genes[[i]]$avg_log2FC < 0, ]
      }
    }
    
    A <- rownames(de_genes[[1]])
    B <- rownames(de_genes[[2]])
    C <- rownames(de_genes[[3]])
    D <- rownames(de_genes[[4]])
    
    # Preparing clusterProfiler to perform hypergeometric test on msigdb signatures 
    msig.gene.set <-
      msigdbr(species = "Homo sapiens", category = "H") %>%
      dplyr::select(gs_name, human_gene_symbol)
    msig.name <-
      msigdbr(species = "Homo sapiens", category = "H") %>%
      dplyr::select(gs_id, gs_name)
    
    # Getting middle "flower" of the Venn diagram
    intersect <- c((intersect(intersect(intersect(A, B), C), D)),
                   (intersect(intersect(A, B), C)),
                   (intersect(intersect(A, B), D)),
                   (intersect(intersect(A, D), C)),
                   (intersect(intersect(B, D), C))) %>%
      unique()
    
    temp <-
      enricher(intersect,
               TERM2GENE = msig.gene.set,
               TERM2NAME = msig.name)@result %>%
      as_tibble() %>%
      select(ID, p.adjust) %>%
      dplyr::rename(pathway = ID) %>%
      arrange(p.adjust) %>%
      slice_head(n = 5) %>%
      mutate(tissue = tissue_name,
             label_y = if_else(direction == "up", 1, -1),
             p.adjust = -log10(p.adjust),
             pathway = str_remove(pathway, "HALLMARK_") %>%
               str_replace_all("_", " "),
             order = seq_len(n()))
    
    if (direction == "down") {
      temp$p.adjust <- temp$p.adjust * -1
    }
    
    fig_2c_data <- bind_rows(fig_2c_data, temp)
  }
}

################ Two sided Barplot
# Properly name and order tissue types
fig_2c_data <- fig_2c_data %>%
  mutate(tissue = stringr::str_replace(tissue,
                                       "Tibia",
                                       "Tibia Gene Sets") %>%
           stringr::str_replace("Lung",
                                "Lung Gene Sets"),
         pathway = stringr::str_wrap(pathway, width = 25))

fig_2c_data$tissue <- factor(as.factor(fig_2c_data$tissue),
                             levels = c("Tibia Gene Sets",
                                        "Lung Gene Sets"))

# Create tibble that is properly ordered to use in proper plotting of tibia and lung data
lab4plot <- tibble(y = c(-4, 4, -4, 4),
                   x = c(0.5, 0.5, 0.5, 0.5),
                   label = as.factor(c("Downregulated\nduring tibia colonization",
                                       "Upregulated\nduring tibia colonization",
                                       "Downregulated\nduring lung colonization",
                                       "Upregulated\nduring lung colonization")),
                   tissue = factor(as.factor(c("Tibia Gene Sets",
                                               "Tibia Gene Sets",
                                               "Lung Gene Sets",
                                               "Lung Gene Sets")),
                                   levels = c("Tibia Gene Sets",
                                              "Lung Gene Sets")))

figure_2c <-
  ggplot() +
  geom_bar(data = fig_2c_data,
           aes(x = -1 * order,
               y = p.adjust,
               fill = p.adjust > 0),
           stat = "identity",
           alpha = 0.8) +
  coord_flip() +
  facet_wrap(~ tissue,
             ncol = 1,
             scales = "free") +
  geom_hline(yintercept = c(-1.5, 1.5),
             color = "gray",
             linetype = 2,
             linewidth = 0.5) +
  geom_hline(yintercept = 0,
             color = "black",
             linewidth = 0.5) +
  geom_text(data = fig_2c_data,
            aes(x = -1 * order,
                y = label_y * 4,
                label = pathway),
            size = 1.5) +
  geom_text(data = lab4plot,
            aes(x = x,
                y = y,
                label = label),
            fontface = "bold",
            size = 2) +
  scale_fill_manual(values = c("#377eb8", "#e41a1c")) +
  theme_bw() +
  theme(strip.background = element_rect(color = "white", fill = "white"),
        strip.text.x = element_text(size = 9, face = "bold"),
        axis.text.x = element_text(size = 4),
        legend.position = "none",
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.border = element_blank(),
        axis.line.x = element_line(color = "black"),
        axis.text.y = element_blank(),
        axis.ticks.y = element_blank()) +
  labs(title = "",
       y = "",
       x = "") +
  ylim(-7, 7) +
  xlim(-5.5, 1)
# Take a moment to ensure reset to null device
dev.off()

figure_2c

# Save as PDF
pdf("Figures/figure_2/figure_2c.pdf",
    width = 6,
    height = 9)
figure_2c
dev.off()

save(figure_2c, file = "Figures/figure_2/figure_2c.RData")
```

## Paneling figure 2

```{r panelFig2, eval = TRUE}
load("Figures/figure_2/figure_2b.RData")
load("Figures/figure_2/figure_2c.RData")

layout_matrix <- matrix(c(1, 1, 2, 2, 5, 5, 5,
                          3, 3, 4, 4, 5, 5, 5),
                        nrow = 2,
                        byrow = TRUE)

pdf("Figures/figure_2/figure_2.pdf",
    width = 6,
    height = 4)
gridExtra::grid.arrange(grobs = list(figure_2b$`Tibia - down`,
                                     figure_2b$`Tibia - up`,
                                     figure_2b$`Lung - down`,
                                     figure_2b$`Lung - up`,
                                     figure_2c),
                        layout_matrix = layout_matrix)
dev.off()
```


# Figure 3

Figure 3. Osteosarcoma cells retain phenotypic heterogeneity despite adaptive changes in response to changing microenvironments. A) Schematic outlining scRNA-seq bioinformatics workflow. In B-D, samples subset to equal number of cells to allow inter-model comparison (n=1500 per condition).

## Figure 3B

B) Osteosarcoma cells exhibited higher ITH scores than that of osteoblasts grown in cell culture (####p <0.001). Within model comparison, identified increase in ITH scores as cells colonize tibia and lung microenvironments, with the exception of NCH-OS-7 (****p <0.001). P values were adjusted using Šidák multiple comparisons test.

```{r Fig3b}
# Load OS data that was subsetted to be equal cell numbers across 
load("Data/OSlist_1500_CCR.RData")
merged_sobjs <-
    SeuratObject:::merge.Seurat(x = OS.list_1500[[1]],
                                y = OS.list_1500[2:length(OS.list_1500)],
                                add.cell.ids = names(OS.list_1500))

merged_sobjs$group <-
    colnames(merged_sobjs) %>%
    str_remove("_.+")

group_names <-
    merged_sobjs$group %>%
    unique()

# Find the genes that are variable within each sample, and combine into a vector
var_genes <- c()
for (i in seq_along(OS.list_1500)) {
  OS.list_1500[[i]] <- OS.list_1500[[i]] %>%
    NormalizeData(verbose = FALSE) %>%
    FindVariableFeatures(verbose = FALSE)
  var_genes <- c(var_genes, VariableFeatures(OS.list_1500[[i]]))
}
var_genes <- unique(var_genes)

# Process each sample to be CC regressed in the same gene space and calculate ITH score
raw_ith <- data.frame(matrix(, nrow = 1500, ncol = 0))
for (i in seq_along(group_names)) {
  better_name <-
    group_names[i] %>%
    str_replace("\\.", " ") %>%
    str_replace("cx", "Culture") %>%
    str_replace("osteoblasts", "Osteoblasts") %>%
    str_replace("t143", "143")

  sub_obj <- subset(merged_sobjs, group == group_names[i]) %>%
    NormalizeData(verbose = FALSE) %>%
    FindVariableFeatures(verbose = FALSE) %>%
    ScaleData(
      features = var_genes,
      vars.to.regress = c("S.Score", "G2M.Score"),
      block.size = 100000)

  z <-
    GetAssayData(sub_obj, slot = "scale.data") %>%
    as.data.frame() %>%
    rownames_to_column("gene") %>%
    as_tibble() %>%
    filter(gene %in% VariableFeatures(sub_obj)) %>%
    select(-gene) %>%
    t()

  mat <- dist(z, diag = TRUE, upper = FALSE)
  mat2 <- as.matrix(mat)
  mat2[upper.tri(mat2, diag = TRUE)] <- NA
  raw_ith[[better_name]] <- colMeans(mat2, na.rm = TRUE)
}

# the last row will be all NAs because we removed the diagonal from the matrix
raw_ith <- na.omit(raw_ith)

ith <- pivot_longer(raw_ith,
                    cols = everything(),
                    names_to = "Sample",
                    values_to = "ITHScore") %>%
    mutate(Sample = as.factor(Sample) %>%
                fct_relevel(group_names %>%
                        str_replace("\\.", " ") %>%
                        str_replace("cx", "Culture") %>%
                        str_replace("osteoblasts", "Osteoblasts") %>%
                        str_replace("t143", "143")))

ith$logITH <- log10(ith$ITHScore)
ridge_cols <- c("#B8B8B8FF",
                "#D43F3AFF", "#D43F3AFF", "#D43F3AFF",
                "#EEA236FF", "#EEA236FF", "#EEA236FF",
                "#357EBDFF", "#357EBDFF", "#357EBDFF",
                "#5CB85CFF", "#5CB85CFF", "#5CB85CFF"
)
rplot <-
  ggplot(ith, aes(x = logITH, y = Sample)) +
  geom_density_ridges(scale = 2, aes(fill = Sample)) +
  scale_y_discrete(expand = c(0, 0), limits = rev) +
  scale_x_continuous(expand = c(0, 0)) +
  scale_fill_manual(values = ridge_cols) +
  theme_ridges() +
  xlab("log[ITH Score]") +
  theme(legend.position = "none",
        axis.text.y = element_text(size = 6),
        axis.text.x = element_text(size = 6),
        axis.title.x = element_text(hjust = 0.5, size = 6),
        axis.title.y = element_blank())

rplot

# Perform a Mann-Whitney U for each group
comps <- tribble(
  ~ctl, ~comp,
  1, 2,
  1, 5,
  1, 8,
  1, 11,
  2, 3,
  2, 4,
  5, 6,
  5, 7,
  8, 9,
  8, 10,
  11, 12,
  11, 13
)
comps$ctl_name <- unique(ith$Sample)[comps$ctl]
comps$comp_name <- unique(ith$Sample)[comps$comp]
comps$mwu_p <- NA
comps$overlap <- NA
comps$eff_size <- NA

# Calculate Mann-Whitney test (p-value), overlap, and effect size (Cohen d)
for (i in seq_along(comps$ctl)) {
  s <- wilcox.test(raw_ith[, comps$ctl[i]], raw_ith[, comps$comp[i]])
  comps$mwu_p[i] <- s$p.value
  s <- overlapping::overlap(list(ctrl = raw_ith[, comps$ctl[i]],
      comp = raw_ith[, comps$comp[i]]))$OV
  comps$overlap[i] <- s
  s <- effsize::cohen.d(raw_ith[, comps$ctl[i]], raw_ith[, comps$comp[i]])
  comps$eff_size[i] <- s$estimate
}

# Convert overlap to percentage
comps$overlap <- paste0(format(comps$overlap * 100, digits = 2, trim = T), "%")

# Plot with p-values superimposed for each sample
figure_3b <-
  rplot +
  geom_text(data = comps,
    aes(x = 1.8, y = comp_name),
    label = paste0("overlap ",
                   comps$overlap,
                   ", eff. size ",
                   format(comps$eff_size, digits = 1, trim = T),
                   "\n"),
    size = 8,
    color = "#2b2a2a")
figure_3b

#Save as PDF
if(!dir.exists("Figures/figure_3")) {
  dir.create("Figures/figure_3")
}

# Save as PDF
pdf("Figures/figure_3/figure_3b.pdf",
    width = 9,
    height = 8)
figure_3b
dev.off()

save(figure_3b, file = "Figures/figure_3/figure_3b.RData")
```

## Figure 3C

C) UMAP analysis for merged primary and metastatic samples in each of the four models. Cells in grey represent remaining cells in merged sample. Cluster enrichment analysis shows distribution of cells in each cluster in the two microenvironment conditions (tibia, lung). While some cells in the primary and metastatic lesions adopted shared phenotypes, others adopted distinct phenotypes.

```{r Fig3c}
load("Data/OS_list_CCR.RData")

OS_list$OS17_TL <- OS_list$OS17_TL %>%
  FindClusters(resolution = 0.2)
OS_list$`143B_TL` <- OS_list$`143B_TL` %>%
  FindClusters(resolution = 0.15)
OS_list$NCHOS2_TL <- OS_list$NCHOS2_TL %>%
  FindClusters(resolution = 0.15)
OS_list$NCHOS7_TL <- OS_list$NCHOS7_TL %>%
  FindClusters(resolution = 0.2)

# Cluster distribution in each sample
# proportion of cells in lung in each cluster
lung_prop_tbl <- tibble()
for (s_obj in c(OS_list$OS17_TL,
                OS_list$`143B_TL`,
                OS_list$NCHOS2_TL,
                OS_list$NCHOS7_TL)) {
  s_obj$src <- factor(s_obj$src,
                      levels = c("Tibia",
                                 "Lung"))
  lung_prop_tbl <-
    s_obj@meta.data %>%
    as_tibble() %>%
    select(model, src, seurat_clusters) %>%
    group_by(model, seurat_clusters) %>%
    summarize(lung_perc = sum(grepl("Lung", src)) / n() * 100,
              .groups = "drop") %>%
    dplyr::rename(sample = model,
                  cluster = seurat_clusters) %>%
    bind_rows(lung_prop_tbl)
}

write_tsv(lung_prop_tbl, "Data/Fig3d_lung_prop_tbl.tsv")

#### UMAP Plots ####
sample_names <- c("OS17_TL",
                  "143B_TL",
                  "NCHOS2_TL",
                  "NCHOS7_TL")

figure_3c_umaps <- list()
for (s_obj in sample_names) {
  # Tibia plot with lung-derived cells greyed out
  temp_obj <- OS_list[[s_obj]]
  lung <- vector(mode = "character")
  temp <- WhichCells(temp_obj,
                     expression = src == "Lung")
  lung <- append(lung, temp)
  Idents(temp_obj, cells = lung) <- " "

  cust_dim_plot <- function(x_obj, title, subtitle) {
      DimPlot(x_obj,
              reduction = "umap",
              pt.size = 0.1,
              label = T,
              label.size = 1,
              cols = c("light grey",
                       "#E64B35FF",
                       "#4DBBD5FF",
                       "#00A087FF",
                       "#3C5488FF"),
              order = T) +
    coord_fixed() +
    NoLegend() +
    NoAxes() +
    ggtitle(title, subtitle = subtitle) +
    theme(plot.title = element_text(hjust = 0.5, size = 7),
          plot.subtitle = element_text(hjust = 0.5, size = 6),
          plot.margin = unit(c(0, 0, 0, 0), "pt"))
  }

  tibia_plot <-
    cust_dim_plot(temp_obj,
                  title = s_obj,
                  subtitle = "Tibia")

  # Lung plot with lung-derived cells greyed out
  temp_obj2 <- OS_list[[s_obj]]
  tib <- vector(mode = "character")
  temp <- WhichCells(temp_obj2,
                     expression = src == "Tibia")
  tib <- append(tib, temp)
  Idents(temp_obj2, cells = tib) <- " "

  lung_plot <-
    cust_dim_plot(temp_obj2,
                  title = "",
                  subtitle = "Lung")

  figure_3c_umaps[[s_obj]] <- tibia_plot + lung_plot


  pdf(paste0("Figures/figure_3/figure_3c1_",
             s_obj,
             ".pdf"),
      width = 5,
      height = 5)
  print(figure_3c_umaps[[s_obj]])
  dev.off()
}

save(figure_3c_umaps, file = "Figures/figure_3/figure_3c_umaps.RData")

#### Bar plots ####
objlist <- c(OS_list$OS17_TL,
             OS_list$`143B_TL`,
             OS_list$NCHOS2_TL,
             OS_list$NCHOS7_TL)

figure_3c_barplots <- list()

for (sobj in objlist) {
  tident <- table(Idents(sobj),
                  sobj$src,
                  dnn = c("cluster",
                          "tissue")) %>%
    as.data.frame() %>%
    group_by(tissue) %>%
    mutate(perc = Freq / sum(Freq) * 100)

  tident$cluster <- as.character(tident$cluster)
  tident$tissue <- factor(tident$tissue, levels = c("Tibia", "Lung"))
  figure_3c_barplots[[sobj$model[1]]] <-
    ggplot(tident,
           aes(x = tissue,
               y = perc,
               fill = cluster)) +
    theme_bw(base_size = 5) +
    geom_col(position = "fill", width = 0.5) +
    xlab("Tissue") +
    ylab("Proportion") +
    scale_fill_npg(alpha = 1) +
    theme_minimal() +
    NoLegend() +
    theme(panel.grid = element_blank(),
          axis.text.y = element_text(size = 5),
          axis.text.x = element_text(size = 5,
                                     angle = 45,
                                     hjust = 1),
          axis.title.x = element_text(size = 6),
          axis.title.y = element_text(size = 6))

  # Save plot as PDF
  pdf(paste0("Figures/figure_3/figure_3c2_",
             head(sobj$model, 1),
             ".pdf"),
      width = 1,
      height = 2)
  print(figure_3c_barplots[[sobj$model[1]]])
  dev.off()
}

save(figure_3c_barplots, file = "Figures/figure_3/figure_3c_barplots.RData")

# For Rmd visual
figure_3c2 <- ggplot(tident,
                     aes(x = tissue,
                         y = perc,
                         fill = cluster)) +
  theme_bw(base_size = 5) +
  geom_col(position = "fill", width = 0.5) +
  xlab("Tissue") +
  ylab("Proportion") +
  scale_fill_npg(alpha = 1)
figure_3c2
```

## Figure 3D

D) Metabolic heterogeneity in glycolysis activation. We used a pathway enrichment
analysis for hallmark gene sets using genes differentially upregulated in each
cluster relative to every other cluster within the same model. P values were
adjusted for multiple comparisons. Boxes in grey identify non-significant pathway
enrichments, whereas boxes in red identify statistically significant enrichments.
Bar plot shows percentage of cells identified per cluster in lung metastases.
In B-D, samples subset to equal number of cells to allow inter-model comparison
(n=1500 per condition).

### Processing

```{r Fig3d_process, dependson='Fig3c'}
library(data.table)
#library(pheatmap)
#Create list of objects used in Figure 3 heatmap
B.list <- list(OS17 = OS_list$`OS17_TL`,
               t143b = OS_list$`143B_TL`,
               OS2 = OS_list$`NCHOS2_TL`,
               OS7 = OS_list$`NCHOS7_TL`)

# Use DGEA function to find hallmark pathway expression
em.hm.list <- list()
set.seed(1337)
for (i in seq_len(length(B.list))) {
  em.hm.list[[i]] <- DGEA(B.list[[i]]) %>%
    setDT(keep.rownames = TRUE) %>%
    as_tibble() %>%
    rename_with(.cols = where(is.numeric), ~ str_c(names(B.list[i]), "_", .x))
}

cx.pd <- inner_join(inner_join(em.hm.list[[1]],
                               em.hm.list[[2]],
                               by = "rn"),
                    inner_join(em.hm.list[[3]],
                               em.hm.list[[4]],
                               by = "rn"),
                    by = "rn")

# Make the "rn" column the rownames -Emily
# cx.pd <- cx.pd %>%
#     column_to_rownames("rn")

head(cx.pd)
##Write table to edit rownames to easily convert to dataframe
write_tsv(cx.pd, file = "Data/cxpd.tsv")
```

### Code I gave to Sanjana at some point for Figure 3d
I modified this from code she gave me to clean it up and get it working

The only data we need are the cs.pd object and the percent of lung cells per cluster
which is entered here as aka4
```{r Fig3d}
lung_percent <-
  read_tsv("Data/Fig3d_lung_prop_tbl.tsv") %>%
  mutate(sample_order = rank(sample) * 1000 - lung_perc,
         x_label = paste0(sample, " c", cluster) %>%
           reorder(sample_order))

cx.pd <-
  read_tsv("Data/cxpd.tsv",
           show_col_types = FALSE) %>%
  pivot_longer(c(-rn),
               names_to = "sample",
               values_to = "pval") %>%
  mutate(cluster = str_remove(sample, ".+_") %>%
           as.numeric(),
         sample = str_replace(sample, "OS2", "NCH-OS-2") %>%
           str_replace("OS7", "NCH-OS-7") %>%
           str_replace("t143b", "143B") %>%
           str_replace("OS17", "OS-17") %>%
           str_remove("_.+"),
         rn = str_remove(rn, "HALLMARK_") %>%
           str_replace("_", " ")) %>%
  left_join(lung_percent) %>%
  mutate(pval = -log10(pval),
         sample_order = rank(sample) * 1000 - lung_perc,
         signif_bool = pval >= (-1 * log10(0.05)),
         signif = if_else(signif_bool == TRUE, "p-value < 0.05", "p-value > 0.05"))

P1 <-
  cx.pd %>%
  ggplot(., aes(x = x_label, y = rn, fill = signif)) +
  geom_tile() +
  geom_text(aes(label = sprintf("%0.2f", pval)), size = 2.2) +
  scale_fill_manual(values = c("red", "gray"),
                    name = "") +
  scale_y_discrete(limits = rev, position = "right") +
  theme(panel.background = element_blank(),
        strip.background = element_blank(),
        strip.text.x = element_blank(),
        plot.margin = unit(c(0, 0, 0, 0), "cm"),
        panel.border = element_rect(colour = "black", fill = NA),
        axis.text.x = element_text(angle = 90,
                                   hjust = 1,
                                   vjust = 0.4,
                                   size = 8),
        axis.text.y = element_text(size = 8),
        legend.background = element_blank(),
        legend.text = element_text(size = 8)) +
  labs(x = "",
       y = "") +
  facet_grid(. ~ sample,
             scales = "free_x",
             space = "free")

P1_legend <- cowplot::get_legend(P1)
P1 <- P1 + theme(legend.position = "none")

P2 <-
  lung_percent %>%
  ggplot(., aes(x = x_label, y = lung_perc, fill = sample)) +
  geom_bar(stat = "identity") +
  ylab("Percent\nin lung") +
  xlab("") +
  labs(fill = "Model",
       y = "Percent in\nlung (%)") +
  scale_fill_brewer(palette = "Set2") +
  theme(axis.text.x = element_blank(),
        axis.ticks.x = element_blank(),
        axis.title.y = element_text(size = 6),
        panel.background = element_blank(),
        strip.background = element_blank(),
        strip.text.x = element_blank(),
        plot.margin = unit(c(0, 0, -0.2, 0), "cm"),
        panel.border = element_rect(colour = "black", fill = NA),
        legend.text = element_text(size = 8)) +
  facet_grid(. ~ sample,
             scales = "free_x",
             space = "free")

P2_legend <- cowplot::get_legend(P2)
P2 <- P2 + theme(legend.position = "none")


figure_3d <-
    cowplot::plot_grid(cowplot::plot_grid(P2,
                                          P1,
                                          align = "v",
                                          rel_heights = c(1, 8),
                                          ncol = 1),
                       cowplot::plot_grid(P1_legend,
                                          P2_legend,
                                          ncol = 1,
                                          axis = "t"),
                        nrow = 1,
                        rel_widths = c(0.95, 0.05))
figure_3d

save(figure_3d, file = "Figures/figure_3/figure_3d.RData")

pdf("Figures/figure_3/figure_3d.pdf",
    width = 15,
    height = 8)
figure_3d
dev.off()
```

```{r assemblingFig3, eval=TRUE}
load("Figures/figure_3/figure_3b.RData")
load("Figures/figure_3/figure_3c_umaps.RData")
load("Figures/figure_3/figure_3c_barplots.RData")
load("Figures/figure_3/figure_3d.RData")

layout_matrix <-
    matrix(c(1,  1,  1,  1,  2,  2,  2,  6,  4,  4,  4,  8,
             1,  1,  1,  1,  2,  2,  2,  6,  4,  4,  4,  8,
             1,  1,  1,  1,  3,  3,  3,  7,  5,  5,  5,  9,
             1,  1,  1,  1,  3,  3,  3,  7,  5,  5,  5,  9,
             10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10,
             10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10,
             10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10,
             10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10,
             10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10,
             10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10,
             10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10),
         nrow = 11,
         byrow = TRUE)

pdf("Figures/figure_3/figure_3.pdf",
    width = 8,
    height = 9)
gridExtra::grid.arrange(grobs = list(figure_3b,
                                     patchwork::patchworkGrob(figure_3c_umaps$OS17_TL),
                                     patchwork::patchworkGrob(figure_3c_umaps$`143B_TL`),
                                     patchwork::patchworkGrob(figure_3c_umaps$NCHOS2_TL),
                                     patchwork::patchworkGrob(figure_3c_umaps$NCHOS7_TL),
                                     figure_3c_barplots$`OS-17`,
                                     figure_3c_barplots$`143B`,
                                     figure_3c_barplots$`NCH-OS-2`,
                                     figure_3c_barplots$`NCH-OS-7`,
                                     figure_3d),
                        layout_matrix = layout_matrix)
dev.off()
```

# Figure 4 - Images

Figure 4. Heterogeneity in Glycolysis activation identified within metastatic osteosarcoma lesions.

Most of figure 4 of this paper is composed of microscopy images and so is not part of the code for this paper. 

Please see OSHetero2021_Supplemental_Figures.Rmd for the patient data code.

# Figure 5

Figure 5. Lung colonization bottleneck enriches for clonal populations with a common transcriptional phenotype. A) Schematic of experimental set-up. B) UMAP analysis of lineage tagged OS-17 cells in culture, implanted into tibia, or inoculated to generate lung metastases. These lineage tags are heritable allowing us to track the transcriptional profiles of daughter cells originating from the same ancestor. Metastatic and orthotopic samples share many cells with similar phenotypes, though cultured cells are dramatically different (n = 2800 per condition) C) Frequency distribution of clones identified in each of the conditions. A high level of clonal diversity is maintained in the orthotopic tumors (78% unique clones in primary tumor compared to 86% unique clones in cell culture); though dominant clones emerge in the lung (only 33% unique clones). D) Overlay of cells sharing a common ancestor onto the dimensional reduction plots in the top four enriched clonal families. E) UMAP visualization of cells belonging to top ten enriched clonal families highlighted in red. F) Pathway enrichment analysis for hallmark gene sets comparing enriched clones relative to the remaining metastatic tumor cells identifies Hypoxia, Glycolysis, EMT and MTORC1 signaling related genes to be significantly upregulated

### Load

```{r Fig5load}
library(rrrSingleCellUtils)

# Load raw objects for lineage tag analysis
load("Data/os17_cx_raw.RData")
load("Data/os17_tib_raw.RData")
load("Data/os17_lung_raw.RData")

# Load merged and cell cycle regressed OS17 object
load("Data/os17.RData")

# Subset all to 2800 cells in each condition
set.seed(108)
os17_cx_raw <- subset(os17_cx_raw,
                      cells = sample(Cells(os17_cx_raw),
                                     2800))

os17_tib_raw <- subset(os17_tib_raw,
                       cells = sample(Cells(os17_tib_raw),
                                      2800))
os17_lung_raw <- subset(os17_lung_raw,
                        cells = sample(Cells(os17_lung_raw),
                                       2800))

# Load lineage tracing data
S0016xS0027_ltbc <- read.delim("Data/S0016xS0027.ltbc")
S0018xS0028_ltbc <- read.delim("Data/S0018xS0028.ltbc")
S0024xS0029_ltbc <- read.delim("Data/S0024xS0029.ltbc")
```

## Figure 5B

B) UMAP analysis of lineage tagged OS-17 cells in culture, implanted into tibia, or inoculated to generate lung metastases. These lineage tags are heritable allowing us to track the transcriptional profiles of daughter cells originating from the same ancestor. Metastatic and orthotopic samples share many cells with similar phenotypes, though cultured cells are dramatically different (n = 2800 per condition)

```{r Fig5b, dependson="Fig5load"}
# Set cell identities to proper cluster resolution
Idents(os17) <- os17$RNA_snn_res.0.3

# Plot os17
set.seed(100)
os17$src <- factor(os17$src, levels = c("Culture", "Tibia", "Lung"))
figure_5b <- DimPlot(os17,
                     reduction = "umap",
                     group.by = "src",
                     pt.size = 1,
                     label = T,
                     label.size = 3,
                     order = FALSE,
                     cols = c("#E64B35FF",
                              "#00A087FF",
                              "#4DBBD5FF")) +
  coord_fixed() +
  ggtitle(" ") +
  #scale_color_npg() +
  theme(axis.text = element_text(size = 6),
        axis.title = element_text(size = 8),
        legend.text = element_text(size = 6),
        legend.title = element_text(size = 8))

figure_5b

if(!dir.exists("Figures/figure_5")){
  dir.create("Figures/figure_5")
}

save(figure_5b, file = "Figures/figure_5/figure_5b.RData")

# Save as PDF
pdf("Figures/figure_5/figure_5b.pdf",
    width = 8,
    height = 4)
figure_5b
dev.off()
```

## Figure 5C

C) Frequency distribution of clones identified in each of the conditions. A high level of clonal diversity is maintained in the orthotopic tumors (78% unique clones in primary tumor compared to 86% unique clones in cell culture); though dominant clones emerge in the lung (only 33% unique clones). 

```{r Fig5c, dependson="Fig5load"}
library(rrrSingleCellUtils)

# There are cells in lung/tibia that cluster with the culture cells and throw off the DimPlots and vise versa

odd_cells <-
  c(subset(os17, subset = src == "Tibia" | src == "Lung") %>%
        Embeddings("umap") %>%
        as.data.frame() %>%
        dplyr::filter(UMAP_1 < 0) %>%
        rownames(),
    subset(os17, subset = src == "Culture") %>%
        Embeddings("umap") %>%
        as.data.frame() %>%
        dplyr::filter(UMAP_1 > 0) %>%
        rownames())


# kick out those two cells
os17 <-
  os17[, !colnames(os17) %in% odd_cells]

dim_plot <- function(x, src_sub, color) {
  sub_sobj <- subset(x, subset = src == src_sub)
  sub_sobj <- subset(sub_sobj, cell = names(is.na(sub_sobj$lt) == FALSE))
  DimPlot(sub_sobj,
          group.by = "src",
          pt.size = 0.1,
          cols = color) +
    theme_void() +
    NoLegend() +
    NoAxes() +
    labs(title = NULL, x = NULL, y = NULL) +
    theme(plot.background = element_rect(colour = "black"))
}

ltbc_hist <- function(df, fill, n_uniq, title) {
  df %>%
    head(n = 40) %>%
    ggplot(aes(x = stats::reorder(Var1, -freq),
                   y = freq)) +
    geom_bar(fill = fill, stat = "identity") +
    theme_bw() +
    theme(plot.title = element_text(size = 8),
          axis.title = element_text(size = 6),
          axis.text = element_text(size = 6),
          axis.text.x = element_blank(),
          axis.ticks.x = element_blank(),
          panel.grid.major = element_blank(),
          panel.grid.minor = element_blank()) +
    labs(x = "",
         y = "Percentage of Cells",
         title = title) +
    ylim(0, 6) +
    annotate("text",
             label = paste(n_uniq,
                           "out of every 1000 cells have a unique clone"),
             x = -Inf,
             y = Inf,
             hjust = -0.4,
             vjust = 3,
             size = 2)
}

n_uniq_1 <-
  (table(os17_cx_raw$lt) == 1) %>%
  sum() / length(unique(na.omit(os17_cx_raw$lt))) * 1000
#789

cells_w_uniq_clones_1 <-
    table(na.omit(os17_cx_raw$lt)) %>%
    as.data.frame() %>%
    summarize(n = sum(Freq == 1) / sum(Freq) * 1000) %>%
    pull(n) %>%
    round()
# 624

# Plot histograms of lineage tracing tags in the Seurat objects
figure_5c1 <-
  process_ltbc(os17_cx_raw,
               cid_lt = S0016xS0027_ltbc,
               histogram = FALSE,
               title = "Culture, Top 40 Clones",
               col_fill = "#E64B35FF",
               relative = TRUE,
               ret_list = TRUE) %>%
  ltbc_hist(fill = "#E64B35FF", cells_w_uniq_clones_1, "Culture, Top 40 Clones")

inset_1 <- dim_plot(os17, "Culture", color = "#E64B35FF")

figure_5c1 <-
  cowplot::ggdraw() +
  cowplot::draw_plot(figure_5c1) +
  cowplot::draw_plot(inset_1, x = 0.7, y = 0.3, width = 0.2, height = 0.7)

# Save as PDF
pdf("Figures/figure_5/figure_5c1.pdf",
    width = 8,
    height = 2)
figure_5c1
dev.off()

n_uniq_2 <-
  (table(os17_tib_raw$lt) == 1) %>%
  sum() / length(unique(na.omit(os17_tib_raw$lt))) * 1000
# 817

cells_w_uniq_clones_2 <-
    table(na.omit(os17_tib_raw$lt)) %>%
    as.data.frame() %>%
    summarize(n = sum(Freq == 1) / sum(Freq) * 1000) %>%
    pull(n) %>%
    round()
# 652

figure_5c2 <- process_ltbc(os17_tib_raw,
                           cid_lt = S0018xS0028_ltbc,
                           histogram = FALSE,
                           title = "Tibia, Top 40 Clones",
                           col_fill = "#00A087FF",
                           relative = TRUE,
                           ret_list = TRUE) %>%
  ltbc_hist(fill = "#00A087FF", cells_w_uniq_clones_2, "Tibia, Top 40 Clones")

inset_2 <- dim_plot(os17, "Tibia", color = "#00A087FF")

figure_5c2 <-
  cowplot::ggdraw() +
  cowplot::draw_plot(figure_5c2) +
  cowplot::draw_plot(inset_2, x = 0.7, y = 0.3, width = 0.2, height = 0.7)

# Save as PDF
pdf("Figures/figure_5/figure_5c2.pdf",
    width = 8,
    height = 2)
figure_5c2
dev.off()

n_uniq_3 <-
  (table(os17_lung_raw$lt) == 1) %>%
  sum() / length(unique(na.omit(os17_lung_raw$lt))) * 1000
# 638

cells_w_uniq_clones_3 <-
    table(na.omit(os17_lung_raw$lt)) %>%
    as.data.frame() %>%
    summarize(n = sum(Freq == 1) / sum(Freq) * 1000) %>%
    pull(n) %>%
    round()
# 243

figure_5c3 <- process_ltbc(os17_lung_raw,
                           cid_lt = S0024xS0029_ltbc,
                           histogram = FALSE,
                           title = "Lung, Top 40 Clones",
                           col_fill = "#4DBBD5FF",
                           relative = TRUE,
                           ret_list = TRUE) %>%
  ltbc_hist(fill = "#4DBBD5FF", cells_w_uniq_clones_3, "Lung, Top 40 Clones")

inset_3 <- dim_plot(os17, "Lung", color = "#4DBBD5FF")
# ^^^ This function fails because there is only one lung cell and only one tibia cell in the os17 dataset at this point

figure_5c3 <-
  cowplot::ggdraw() +
  cowplot::draw_plot(figure_5c3) +
  cowplot::draw_plot(inset_3, x = 0.7, y = 0.3, width = 0.2, height = 0.7)

# Save as PDF
pdf("Figures/figure_5/figure_5c3.pdf",
    width = 8,
    height = 2)
figure_5c3
dev.off()

figure_5c <-
  gridExtra::grid.arrange(figure_5c1,
                          figure_5c2,
                          figure_5c3,
                          ncol = 1)

save(figure_5c, file = "Figures/figure_5/figure_5c.RData")
```

## Figure 5D

D) Overlay of cells sharing a common ancestor onto the dimensional reduction plots in the top four enriched clonal families.

### Processing

```{r Fig5d_processing, dependson=c("Fig5load","Fig5b","Fig5c")}
# Extract the barcode frequency lists

cx_lt_list <- process_ltbc(os17_cx_raw,
                           cid_lt = S0016xS0027_ltbc,
                           ret_list = TRUE)

tib_lt_list <- process_ltbc(os17_tib_raw,
                           cid_lt = S0018xS0028_ltbc,
                           ret_list = TRUE)

lung_lt_list <- process_ltbc(os17_lung_raw,
                            cid_lt = S0024xS0029_ltbc,
                            ret_list = TRUE)

# Subset the three samples, but retain the current umap data
cx_sub <- subset(os17, subset = src == "Culture")
tib_sub <- subset(os17, subset = src == "Tibia")
lung_sub <- subset(os17, subset = src == "Lung")

# Subset to cells that have a lineage tag
# Culture
Culture.LTcells <- vector(mode = "character")
for (i in 1:nrow(cx_lt_list)) {
    temp <- WhichCells(cx_sub,
                        expression = lt == cx_lt_list[[i, 1]])
    Culture.LTcells <- append(Culture.LTcells,
                                temp)
}

cx.tagged <- subset(cx_sub,
                    cells = Culture.LTcells)

tib.LTcells <- vector(mode = "character")
for (i in 1:nrow(tib_lt_list)) {
  temp <- WhichCells(tib_sub,
                     expression = lt == tib_lt_list[[i, 1]])
  tib.LTcells <- append(tib.LTcells, temp)
}
tib.tagged <- subset(tib_sub,
                     cells = tib.LTcells)

lung.LTcells <- vector(mode = "character")
for (i in 1:nrow(lung_lt_list)) {
  temp <- WhichCells(lung_sub,
                     expression = lt == lung_lt_list[[i, 1]])
  lung.LTcells <- append(lung.LTcells,
                         temp)
}
lung.tagged <- subset(lung_sub,
                      cells = lung.LTcells)
```


### Make enriched clone DimPlots
```{r Fig5d, dependson=c("Fig5load","Fig5b","Fig5c","Fig5d_processing")}
dataset_list <- list(cx.tagged,
                     tib.tagged,
                     lung.tagged)

lt_list <- list(cx_lt_list,
                tib_lt_list,
                lung_lt_list)

sub_list <- list(cx_sub,
                 tib_sub,
                 lung_sub)

color_list <- list("#E64B35FF",
                   "#00A087FF",
                   "#4DBBD5FF")

figure_5d_list <- list()
plot_num <- 1

for (i in seq_len(length(dataset_list))) {
  dataset <- dataset_list[[i]]
  lt_data <- lt_list[[i]]
  sub_data <- sub_list[[i]]
  data_color <- color_list[[i]]
  for (j in 1:4) {
    figure_5d_list[[plot_num]] <-
      patchwork::patchworkGrob(DimPlot(dataset,
                                       reduction = "umap",
                                       pt.size = 0.1,
                                       cells.highlight = WhichCells(sub_data,
                                                                    expression = lt == lt_data[[j, 1]]),
                                       cols.highlight = data_color,
                                       sizes.highlight = 1) +
                                 coord_fixed() +
                                 theme(legend.position = "none") +
                                 ggtitle(paste("Clone",
                                               lt_data[[j, 1]])) +
                                 NoLegend() +
                                 NoAxes() +
                                 theme(title = element_text(size = 6),
                                       plot.margin = unit(c(0, 0, 0, 0), "cm"),
                                       plot.title = element_text(hjust = 0.5)))
    plot_num <- plot_num + 1
  }
}

figure_5d <-
  gridExtra::grid.arrange(grobs = figure_5d_list, ncol = 4)

save(figure_5d, file = "Figures/figure_5/figure_5d.RData")
```

## Figure 5E

E) UMAP visualization of cells belonging to top ten enriched clonal families highlighted in red.

```{r Fig5e, dependson=c("Fig5load","Fig5b","Fig5c","Fig5d_processing")}
# Compare Enriched vs non-enriched 
# Group first 10 against the tagged Lung
LT <- vector(mode = "character")

for (i in 1:10) {
  temp <- WhichCells(lung.tagged,
                     expression = lt == lung_lt_list[[i, 1]])
  LT <- append(LT, temp)
}

Idents(lung.tagged, cells = LT) <- "Enriched clones"

cell.ids <- sample(colnames(lung.tagged))

# Culture plots
figure_5e <- DimPlot(lung_sub,
                     reduction = "umap",
                     pt.size = 1,
                     cells.highlight = WhichCells(lung.tagged,
                                                  idents = "Enriched clones"),
                     cols.highlight = "#E64B35FF",
                     sizes.highlight = 2,
                     order = cell.ids) +
  coord_fixed() + 
  scale_color_discrete(name = "Clones",
                       type = c("#C3C3C3",
                                "#E64B35FF"),
                       labels = c("Unselected" = "Non-enriched clones",
                                  "Group_1" = "Enriched clones")) +
  theme(axis.text = element_text(size = 6),
        axis.title = element_text(size = 8),
        legend.text = element_text(size = 6),
        legend.title = element_text(size = 8),
        legend.position = c(0.9, 0.2),
        legend.background = element_blank(),
        plot.margin = unit(c(0, 0, 0, 0), "cm"))
figure_5e

save(figure_5e, file = "Figures/figure_5/figure_5e.RData")

pdf("Figures/figure_5/figure_5e.pdf",
    width = 6,
    height = 5)
figure_5e
dev.off()
```

## Figure 5F

F) Pathway enrichment analysis for hallmark gene sets comparing enriched clones relative to the remaining metastatic tumor cells identifies Hypoxia, Glycolysis, EMT and MTORC1 signaling related genes to be significantly upregulated

```{r Fig5f, dependson=c("Fig5load","Fig5b","Fig5c","Fig5d_processing","Fig5e")}
# Pathways associated with Enriched clonal families
source("Downstream.v2.R")

# Rename identities from the total lung tumor to include enriched clones
lung_sub_enr <- SetIdent(lung_sub,
                         cells = WhichCells(lung.tagged,
                                            idents = "Enriched clones"),
                         value = "Enriched clones")

# Analyze enriched clones against the remaining tumor cells in the lung
lung_sub_enr <- RenameIdents(lung_sub_enr,
                             "0" = "Remaining Tumor Cells",
                             "1" = "Remaining Tumor Cells",
                             "3" = "Remaining Tumor Cells",
                             "6" = "Remaining Tumor Cells")

# Save for use in supplement
save(lung_sub_enr, file = "Data/lung_sub_enr.RData")

# Create object list for DGEA hallmark analysis
B.list <- list(OS17 = lung_sub_enr)

# Perform differential gene expression analysis for hallmark pathways
set.seed(100)
em.hm.list <- list()
for (i in seq_along(B.list)) {
  em.hm.list[[i]] <- DGEA(B.list[[i]])
}

# Achieve -log10 of the p-value
temp1 <- em.hm.list[[1]]
temp1 <- -log10(temp1)

# Remove unnecessary characters
c <- rownames(temp1)
c <- gsub("_", " ", c)
c <- gsub("HALLMARK ", "", c) 
rownames(temp1) <- c

# Remove negative 0 values due to number of decimal points displayed
temp1 <- abs(temp1)
temp1 <- temp1["Enriched clones"]

# Subset to only pathways with significant pvalue (-1 * log10(0.05)) = 1.30103)
temp1["Enriched clones"] >= 1.30103 -> temp1$logical
colnames(temp1) <- c("pvalue",
                     "significant")
logical <- temp1$significant
temp1 <- temp1[logical, ]
temp1 <- rownames_to_column(temp1,
                            var = "Pathway")

# Order by -log10pvalue for plotting
temp1 <-
  temp1 %>%
  mutate(Pathway = str_wrap(Pathway, width = 30) %>%
           as.factor() %>%
           fct_reorder(pvalue))

# Plot 
figure_5f <- ggplot(temp1,
                    aes(x = pvalue,
                        y = Pathway)) +
  geom_bar(stat = "identity") +
  theme_bw() +
  ylab("") +
  xlab("-log10(p-value)") +
  geom_vline(xintercept = 1.5,
             color = "gray",
             linetype = 2,
             linewidth = 1) +
  theme(axis.text = element_text(size = 6),
        axis.title = element_text(size = 8))

figure_5f

save(figure_5f, file = "Figures/figure_5/figure_5f.RData")

pdf("Figures/figure_5/figure_5f.pdf",
    width = 8,
    height = 3)
figure_5f
dev.off()
```

### Test if clones tend to be enriched within a single cluster
This was ultimately put into Emily's code for the supplementary Figures
```{r eval=FALSE}
load("Data/os17.RData")

n_clones <- 30
top_clones <-
  tibble(src = os17$src,
         clone = os17$lt,
         cluster = paste0("c_", os17$seurat_clusters)) %>%
  na.omit() %>%
  group_by(src, clone, cluster) %>%
  mutate(n_cells_in_cluster = n()) %>%
  ungroup() %>%
  group_by(src, clone) %>%
  mutate(n_cells_in_clone = n()) %>%
  ungroup() %>%
  group_by(src) %>%
  arrange(desc(n_cells_in_clone), clone, .by_group = TRUE) %>%
  unique() %>%
  pivot_wider(id_cols = c(src, clone, n_cells_in_clone),
              names_from = cluster,
              values_from = n_cells_in_cluster) %>%
  filter(n_cells_in_clone >= 5) %>%
  #slice_head(n = n_clones) %>%
  ungroup() %>%
  pivot_longer(cols = starts_with("c_"),
               names_to = "cluster",
               values_to = "n_cells_in_cluster") %>%
  na.omit() %>%
  mutate(clone = fct_reorder(clone, n_cells_in_clone))

cell_counts_to_test <-
  top_clones$n_cells_in_clone %>%
  unique() %>%
  sort()

names(cell_counts_to_test) <- paste0("c_", cell_counts_to_test)

# Get the number of clusters represented by n randomly selected cells
max_per_cluster_n_cells <- function(sobj, n_cells) {
  selected_cell_indexes <-
    sample(seq_len(ncol(sobj)),
           size = n_cells,
           replace = FALSE)
  
  max_n_per_cluster <-
    sobj$seurat_clusters[selected_cell_indexes] %>%
    table() %>%
    max()
  return(max_n_per_cluster)
}

# bootstrap the data n_perms times to get a distribution
perm_n_times <- function(sobj, n_cells, n_perms = 1000) {
  message("Testing ", n_cells, " cells")
  n_clusters <- replicate(n_perms, max_per_cluster_n_cells(sobj, n_cells))
  return(n_clusters)
}

# Do this for all observed cells/clone
n_perms <- 100000
perm_table <-
  sapply(cell_counts_to_test,
         function(cell_num) perm_n_times(os17,
                                         cell_num,
                                         n_perms = n_perms),
         USE.NAMES = TRUE) %>%
  as.data.frame()

top_clones %>%
  rowwise() %>%
  mutate(p_value = (
    sum(perm_table[[paste0("c_", n_cells_in_clone)]] >=
          n_cells_in_cluster) /
      n_perms
  ),
  signif = p_value <= 0.05,
  cluster = str_remove(cluster, "c_")) %>%
  ggplot(aes(x = cluster,
             y = clone,
             fill = n_cells_in_cluster)) +
  geom_tile() +
  geom_text(aes(label = n_cells_in_cluster,
                color = signif)) +
  scale_color_manual(values = c("gray", "red")) +
  facet_grid(src ~ ., scales = "free_y", space = "free") +
  labs(x = "Cluster",
       y = "Clone",
       fill = "Cells in Cluster",
       color = "Enriched Cluster") +
  theme_bw()

```


### Panel Figure 5
```{r Fig5_panel, dependson=c("Fig5load","Fig5b","Fig5c","Fig5d_processing","Fig5e","Fig5f")}
load("Figures/figure_5/figure_5b.RData")
load("Figures/figure_5/figure_5c.RData")
load("Figures/figure_5/figure_5d.RData")
load("Figures/figure_5/figure_5e.RData")
load("Figures/figure_5/figure_5f.RData")

layout_matrix <- matrix(c(NA, 1, 2, 3, 4, 5),
                        ncol = 2,
                        byrow = TRUE)

pdf("Figures/figure_5/figure_5.pdf",
    width = 8,
    height = 11)
gridExtra::grid.arrange(grobs = list(figure_5b,
                                     figure_5c,
                                     figure_5d,
                                     figure_5e,
                                     figure_5f),
                        layout_matrix = layout_matrix,
                        ncol = 2,
                        nrow = 3)
dev.off()
```


```{r session}
sessionInfo()
```
