---
title: "Updated OSHetero2021 Supplement"
author: "Sanjana Rajan, Emily Franz, Matthew Cannon"
date: "`r format(Sys.time(), '%m/%d/%Y')`"
output: 
  html_document:
    toc: true
    toc_float: true
    toc_depth: 5
    number_sections: false
    code_folding: hide
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
  tidy = TRUE,
  echo = TRUE,
  cache = TRUE,
  collapse = TRUE,
  tidy.opts = list(width.cutoff = 95),
  message = FALSE,
  warning = FALSE,
  cache.lazy = FALSE,
  fig.show='hide'
)
```

```{r lib, cache=FALSE}
#source("scSeurat.R")
source("Downstream.v2.R")
set.seed(888)
# loading libraries
library(Seurat)
library(future)
library(ggplot2)
library(msigdbr)
library(clusterProfiler)
library(SingleR)
library(dplyr)
library(pheatmap)
library(RColorBrewer)
library(viridis) # inferno color palette
library(grid)
library(ggsci)
library(data.table)
library(ggvenn)
library(tibble)
library(tidyr)
library(tidyverse)
library(stringr)
library(reshape2)
library(pzfx)
library(ggridges)
library(effsize)
library(perm)
library(patchwork)

options(future.globals.maxSize= 89128960000)
```

Note: supplemental numbering within the code chunks reflects the previous version of the paper submitted, visible on bioRx. Edits on supplemental numbering was previously made in Illustrator. 

# S1

```{r s1}
library(rrrSingleCellUtils)

# Prior to cell-cycle regression
load("Data/os17_cx_raw.RData")

os17_cx_noccr <- NormalizeData(os17_cx_raw) %>%
  FindVariableFeatures(selection.method = "vst") %>%
  ScaleData() %>%
  RunPCA(pc.genes = os17_cx_raw@var.genes, npcs = 20) %>%
  RunUMAP(reduction = "pca", dims = 1:20) %>%
  FindNeighbors(reduction = "pca", dims = 1:20) %>%
  FindClusters(resolution = 0.3)

os17_cx_noccr <- kill_cc(os17_cx_noccr,
                         cc_regress = "N",
                         find_pcs = 20,
                         use_pcs = 3,
                         use_res = 0.3,
                         method = "umap")
os17_cx_noccr$Phase <- factor(os17_cx_noccr$Phase,
                              levels = c("G1",
                                         "G2M",
                                         "S"))
# UMAP
sfig_s1_a1 <- DimPlot(os17_cx_noccr,
                      group.by = "Phase",
                      label = T) +
  scale_color_npg(alpha = 1) +
  theme(aspect.ratio = 1) +
  ggtitle("")
sfig_s1_a1

# Save as pdf
if(!dir.exists("Figures/supplement")) {
  dir.create("Figures/supplement")
}
if(!dir.exists("Figures/supplement/s1")) {
  dir.create("Figures/supplement/s1")
}
pdf("Figures/supplement/s1/sfig_s1_a1.pdf",
    height = 5,
    width = 5)
sfig_s1_a1
dev.off()

# Post-cell cycle regression
load("Data/os17_cx_CCR.RData")
os17_cx$Phase <- factor(os17_cx$Phase,
                        levels = c("G1",
                                   "G2M",
                                   "S"))
# UMAP
sfig_s1_a2 <- DimPlot(os17_cx,
                      group.by = "Phase",
                      label = F) +
  scale_color_npg(alpha = 1) +
  theme(aspect.ratio = 1) +
  ggtitle("")
sfig_s1_a2

pdf("Figures/supplement/s1/sfig_s1_a2.pdf",
    height = 5,
    width = 5)
sfig_s1_a2
dev.off()

# PCA
# Prior to cell-cycle regression
sfig_s1_b1 <- DimPlot(os17_cx_noccr,
                      group.by = "Phase",
                      label = T,
                      reduction = "pca") +
  scale_color_npg(alpha = 1) +
  theme(aspect.ratio = 1) +
  ggtitle("")
sfig_s1_b1

pdf("Figures/supplement/s1/sfig_s1_b1.pdf",
    height = 5,
    width = 5)
sfig_s1_b1
dev.off()

# PCA
# Post-cell cycle regression
os17_cx$Phase <- factor(os17_cx$Phase,
                        levels = c("G1",
                                   "G2M",
                                   "S"))
sfig_s1_b2 <- DimPlot(os17_cx,
                      group.by = "Phase",
                      label = F,
                      reduction = "pca") +
  scale_color_npg(alpha = 1) +
  theme(aspect.ratio = 1) +
  ggtitle("")
sfig_s1_b2

pdf("Figures/supplement/s1/sfig_s1_b2.pdf",
    height = 5,
    width = 5)
sfig_s1_b2
dev.off()

# Cell-cycle genes Pre-CCR
sfig_s1_c1 <- RidgePlot(os17_cx_noccr,
                        features = c("PCNA",
                                     "TOP2A",
                                     "MCM6",
                                     "MKI67"),
                        group.by = "Phase",
                        ncol = 2,
                        same.y.lims = T,
                        combine = T)
sfig_s1_c1
pdf("Figures/supplement/s1/sfig_s1_c1.pdf",
    height = 4,
    width = 8)
sfig_s1_c1
dev.off()


# Cell-cycle genes - post-CCR
sfig_s1_c2 <- RidgePlot(os17_cx,
                        features = c("PCNA",
                                     "TOP2A",
                                     "MCM6",
                                     "MKI67"),
                        group.by = "Phase",
                        ncol = 2,
                        same.y.lims = T,
                        combine = T,
                        slot = "scale.data")
sfig_s1_c2
pdf("Figures/supplement/s1/sfig_s1_c2.pdf",
    height = 4,
    width = 8)
sfig_s1_c2
dev.off()
```

```{r s1_group, dependson="s1"}

s1 <- (sfig_s1_a1 | sfig_s1_a2) /
  (sfig_s1_b1 | sfig_s1_b2) /
  (sfig_s1_c1 | sfig_s1_c2)
s1
pdf("Figures/supplement/s1/s1.pdf",
    height = 12,
    width = 12)
s1
dev.off()

s1_v2 <- (sfig_s1_a1 + sfig_s1_b1 + sfig_s1_c1) /
  (sfig_s1_a2 + sfig_s1_b2 + sfig_s1_c2)
s1_v2
pdf("Figures/supplement/s1/s1_v2.pdf",
    height = 8,
    width = 15)
s1_v2
dev.off()
```

# S5
(old S2)

```{r s5}
load("Data/osteoblasts.RData")
set.seed(100)

osteoblasts <- NormalizeData(osteoblasts) %>%
  FindVariableFeatures(selection.method = "vst") %>% 
  ScaleData() %>%
  RunPCA(pc.genes = osteoblasts@var.genes, npcs = 20) %>%
  RunUMAP(reduction = "pca", dims = 1:20) %>%
  FindNeighbors(reduction = "pca", dims = 1:20) %>%
  FindClusters(resolution = 0.1)

s.genes <- cc.genes$s.genes
g2m.genes <- cc.genes$g2m.genes
osteoblasts <- CellCycleScoring(object = osteoblasts,
                                s.features = s.genes,
                                g2m.features = g2m.genes,
                                set.ident = TRUE)

# Prior to cell cycle regression
DimPlot(osteoblasts, reduction = "umap",
        group.by = "Phase",
        pt.size = 1) +
  coord_fixed()

# Regress out cell cycle genes
osteoblasts_ccr <- ScaleData(object = osteoblasts,
                             vars.to.regress = c("S.Score",
                                                 "G2M.Score"),
                             features = rownames(x = osteoblasts))

# Save
if(!dir.exists("Data/Supplement")) {
  dir.create("Data/Supplement")
}
save(osteoblasts_ccr, file = "Data/Supplement/osteoblasts_ccr.RData")
```

```{r obplot, dependson="s5"}
#### S2 A1: Osteoblasts UMAP ####
osteoblasts_ccr <- RunPCA(osteoblasts_ccr,
                          pc.genes = osteoblasts_ccr@var.genes,
                          npcs = 20) %>%
  RunUMAP(reduction = "pca", dims = 1:20) %>%
  FindNeighbors(reduction = "pca", dims = 1:20) %>%
  FindClusters(resolution = 0.1)

osteoblasts.nC <- nRes(osteoblasts_ccr,
                       res = seq(from = 0.04,
                                 to = 0.5,
                                 by = 0.01))
pSil(osteoblasts.nC, 0.07)
pSil(osteoblasts.nC, 0.06)
osteoblasts_ccr <- osteoblasts_ccr %>%
  FindClusters(resolution = 0.05)

sfig_s2_a1 <- DimPlot(osteoblasts_ccr,
                      label = T) +
  ggtitle("Osteoblasts") +
  NoLegend() +
  NoAxes() +
  scale_color_npg()

# Save as pdf
if(!dir.exists("Figures/supplement/s2")) {
  dir.create("Figures/supplement/s2")
}
pdf("Figures/supplement/s2/sfig_s2_a1.pdf",
    height = 5,
    width = 5)
sfig_s2_a1
dev.off()

#### S2 A2: Osteoblasts Cell Cycle Pie ####
cid <- sort(unique(osteoblasts_ccr@active.ident))

vplayout <- function(x, y) viewport(layout.pos.row = x, layout.pos.col = y)
pdf("Figures/supplement/s2/sfig_s2_a2_osteoblasts_pie.pdf",
    width = 3,
    height = 2)
grid.newpage()
pushViewport(viewport(layout = grid.layout(1, length(unique(osteoblasts_ccr@active.ident)))))

pies <- list()
for (i in 1:length(unique(osteoblasts_ccr@active.ident))) {
  temp <- WhichCells(osteoblasts_ccr,
                     idents = cid[[i]])
  LT.cells <- subset(osteoblasts_ccr,
                     cells = temp)
  df <- table(LT.cells$Phase)
  bp <- ggplot(as.data.frame(df),
               aes(x = "",
                   y = Freq,
                   fill = Var1)) +
    geom_bar(width = 1,
             stat = "identity")
  pie <- bp +
    coord_polar("y",
                start = 0) +
    scale_fill_brewer(palette = "Blues") +
    theme_minimal() +
    NoAxes() +
    NoLegend() +
    ggtitle(paste("Cluster",
                  cid[[i]]))
  
  pies[[i]] <- pie
  print(pie, vp = vplayout(ceiling(i/length(unique(osteoblasts_ccr@active.ident))), i))
}
dev.off()


#### S2 B Heatmap of osteoblasts markers ####

## Pathways associated with osteoblasts clusters
set.seed(108)
source("Downstream.v3.R")

em.hm.list <- DGEA(osteoblasts_ccr,
                   category = "C2",
                   subcategory = "CP")

temp1 <- -log10(em.hm.list)
# remove '_' by removing special characters
c <- rownames(temp1)
c <- gsub("_", " ", c)
c <- gsub("SIG ", "", c)
c <- gsub("NABA ", "", c)
c <- gsub("SA ", "", c)
rownames(temp1) <- c
# remove negative 0 values due to number of decimal points displayed
temp1 <- abs(temp1)

sfig_s2_b <- temp1 %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column(var = "Sample") %>%
  pivot_longer(c(-Sample),
               names_to = "Pathway",
               values_to = "pval",
               names_repair = "minimal") %>%
  mutate(Signif = pval >= (-1 * log10(0.05))) %>%
  ggplot(., aes(x = Sample,
                y = Pathway,
                fill = Signif)) +
  geom_tile() +
  geom_text(aes(label = sprintf("%0.2f",
                                pval))) +
  scale_fill_manual(values = c("gray",
                               "red"))  +
  scale_y_discrete(limits = rev,
                   position = "right") +
  theme(legend.position = "none",
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        axis.ticks.x = element_blank()) +
  ylab("")

# View in Rmd output
sfig_s2_b

# Save plot as PDF
pdf("Figures/supplement/s2/sfig_s2_b_osteoblasts_Pathway.pdf",
    height = 6,
    width = 6)
plot(sfig_s2_b)
dev.off()
```

```{r s5_group, dependson=c("s5","obplot")}
layout <- "
AAAA#DD
AAAA#DD
BBCC#DD"
s2 <- (sfig_s2_a1 + theme(aspect.ratio = 1)) +
  pies[[1]] +
  pies[[2]] +
  sfig_s2_b +
  plot_layout(design = layout)
s2
pdf("Figures/supplement/s2/s2.pdf",
    height = 7,
    width = 11)
s2
dev.off()
```

# S7
(old S3)

```{r s7}
# Load merged OS models
load("Data/OS_merged_postCCR.RData")

# Rename ob
OS$src[grep("Osteoblasts", OS$src)] <- "Culture"

OS$model <- factor(as.factor(OS$model),
                   levels = c("Osteoblasts",
                              "OS-17",
                              "143B",
                              "NCH-OS-2",
                              "NCH-OS-7"))

#### S3 A ####
# Plot the data
sfig_s3_a1 <- DimPlot(OS,
                      reduction = "umap",
                      group.by = "src",
                      pt.size = 1,
                      label = F) +
  theme(aspect.ratio = 1) +
  ggtitle("OS by Tissue Source") +
  scale_color_npg(alpha = 1)
sfig_s3_a1

# Save as pdf
if(!dir.exists("Figures/supplement/s3")) {
  dir.create("Figures/supplement/s3")
}
pdf("Figures/supplement/s3/sfig_s3_a1.pdf",
    height = 5,
    width = 5)
sfig_s3_a1
dev.off()

sfig_s3_a2 <- DimPlot(OS,
                      reduction = "umap",
                      group.by = "model",
                      pt.size = 1,
                      label = F,
                      cols = c("#828282",
                               "#E64B35FF",
                               "#4DBBD5FF",
                               "#00A087FF",
                               "#3C5488FF"))  +
  theme(aspect.ratio = 1) +
  ggtitle("OS by Model") 
sfig_s3_a2

pdf("Figures/supplement/s3/sfig_s3_a2.pdf",
    height = 5,
    width = 5)
sfig_s3_a2
dev.off()

#### S3 B: Harmony ####
set.seed(108)
library("harmony")

# Run Harmony
OS_harmony <- OS %>%
  RunPCA(pc.genes = OS@var.genes, npcs = 20) %>%
  RunHarmony(group.by.vars = "model", plot_convergence = T) %>%
  RunUMAP(reduction = "harmony", dims = 1:20) %>%
  FindNeighbors(reduction = "harmony", dims = 1:20) %>%
  FindClusters(resolution = 0.2)

# Plot the data
sfig_s3_b <- DimPlot(OS_harmony,
                     reduction = "umap",
                     group.by = "model",
                     pt.size = 1,
                     label = F,
                     shuffle = T,
                     cols = c("#828282",
                              "#E64B35FF",
                              "#4DBBD5FF",
                              "#00A087FF",
                              "#3C5488FF"))  +
  theme(aspect.ratio = 1) +
  ggtitle("OS by Model") 
sfig_s3_b

pdf("Figures/supplement/s3/sfig_s3_b.pdf",
    height = 5,
    width = 5)
sfig_s3_b
dev.off()
```

```{r s7_group, dependson="s7"}
s3 <- (sfig_s3_a2 | sfig_s3_b)
s3

pdf("Figures/supplement/s3/s3.pdf",
    height = 5,
    width = 10)
s3
dev.off()

#CLEAN
# BatchEffectCorrectionSCTransform - all
rm(list=setdiff(ls(), "tenx_load_qc"))
options(future.globals.maxSize= 8912896000000)
```

# S8
(old S4)

## Batch Correction Test
### S8A

```{r s8}
# Determine overlay of replicates with or without the use
# of batch-effect correction methods such as Harmony, FastMNN

#source("scSeurat.R")
source("Downstream.v2.R")
set.seed(108)
# loading libraries
library(Seurat)
library(future)
library(tidyverse)
library(msigdbr)
library(harmony)
library(SeuratWrappers)
library(rrrSingleCellUtils)

# Cellecta - Replicate 1 Load cell culture data
cellecta.cx.raw <- tenx_load_qc("/gpfs0/home2/gdrobertslab/lab/Counts/S0015-cellecta/outs/filtered_feature_bc_matrix/",
                              species_pattern = "^hg19-")
cellecta.cx.raw <- subset(cellecta.cx.raw,
                          subset = nFeature_RNA > 2500 &
                            nCount_RNA < 40000 &
                            percent.mt < 14)
cellecta.cx.raw$src <- "Cellecta_Cx"
cellecta.cx.raw$cond <- "Plate"
cellecta.cx.raw$rep <- "Rep2"

# Load orthotopic lung data
cellecta.lung.raw <- tenx_load_qc("/gpfs0/home2/gdrobertslab/lab/Counts/S0023-cellecta-lung/outs/filtered_feature_bc_matrix/",
                                species_pattern = "^hg19-")
cellecta.lung.raw <- subset(cellecta.lung.raw,
                            subset = percent.mt < 20 &
                              percent.mt > 2)
cellecta.lung.raw$src <- "Cellecta_Lung"
cellecta.lung.raw$cond <- "Lung"
cellecta.lung.raw$rep <- "Rep2"

# Zymo - Replicate 2
zymo.cx.raw <- tenx_load_qc(
  path_10x = "/gpfs0/home/gdrobertslab/lab/Counts/S0016xS0027/outs/filtered_feature_bc_matrix/",
  species_pattern = "^hg19-"
)
zymo.cx.raw <- subset(zymo.cx.raw,
                      subset = nFeature_RNA > 3500 &
                        nCount_RNA < 100000 &
                        percent.mt < 15)
zymo.cx.raw$src <- "Zymo_Cx"
zymo.cx.raw$cond <- "Plate"
zymo.cx.raw$rep <- "Rep1"

# Load orthotopic bone data
zymo.tib.raw <- tenx_load_qc("/gpfs0/home2/gdrobertslab/lab/Counts/S0018xS0028/outs/filtered_feature_bc_matrix/",
                             species_pattern = "^hg19-")
zymo.tib.raw <- subset(zymo.tib.raw,
                       subset = nFeature_RNA > 3000 &
                         nCount_RNA < 60000 &
                         percent.mt < 18)
zymo.tib.raw$src <- "Zymo_Tib"
zymo.tib.raw$cond <- "Bone"
zymo.tib.raw$rep <- "Rep1"

# Load orthotopic lung data
zymo.lung.raw <- tenx_load_qc("/gpfs0/home2/gdrobertslab/lab/Counts/S0024xS0029/outs/filtered_feature_bc_matrix/",
                            species_pattern = "^hg19-")
zymo.lung.raw <- subset(zymo.lung.raw,
                        subset = nFeature_RNA > 1250 &
                          nCount_RNA < 60000 &
                          percent.mt < 25)
zymo.lung.raw$src <- "Zymo_Lung"
zymo.lung.raw$cond <- "Lung"
zymo.lung.raw$rep <- "Rep1"

# Subset to least number of cells across all conditions
# (Cell culture and orthotopic lung)
# Subset all to #1900 cells

cellecta.cx.raw <- subset(cellecta.cx.raw,
                          cells = sample(Cells(cellecta.cx.raw),
                                         1900))
cellecta.lung.raw <- subset(cellecta.lung.raw,
                            cells = sample(Cells(cellecta.lung.raw),
                                           1900))
zymo.cx.raw <- subset(zymo.cx.raw,
                      cells = sample(Cells(zymo.cx.raw),
                                     1900))
zymo.lung.raw <- subset(zymo.lung.raw,
                        cells = sample(Cells(zymo.lung.raw),
                                       1900))

# Merge raw count matrices of these four datsets into a single Seurat object
Rep1_H <- merge(cellecta.cx.raw,
                y = c(cellecta.lung.raw,
                      zymo.cx.raw,
                      zymo.lung.raw),
                add.cell.ids = c("Cellecta_Cx",
                                 "Cellecta_Lung",
                                 "Zymo_Cx",
                                 "Zymo_Lung"), 
                project = "Heterogeneity")

# Dimensionality reduction and clustering
Rep1_H <- NormalizeData(Rep1_H) %>%
  FindVariableFeatures(selection.method = "vst") %>%
  ScaleData() %>%
  RunPCA(pc.genes = Rep1_H@var.genes, npcs = 20) %>%
  RunUMAP(reduction = "pca", dims = 1:20) %>%
  FindNeighbors(reduction = "pca", dims = 1:20) %>%
  FindClusters(resolution = 0.2)  ##Note change in resolution

# Determine contribution of cell cycle related genes to variation in the datasets
s.genes <- cc.genes$s.genes
g2m.genes <- cc.genes$g2m.genes
Rep1_H <- CellCycleScoring(object = Rep1_H,
                           s.features = s.genes,
                           g2m.features = g2m.genes,
                           set.ident = TRUE)
DimPlot(Rep1_H,
        reduction = "umap",
        pt.size = 1,
        label = F,
        group.by = "Phase") +
  coord_fixed() +
  ggtitle("Reps by Source") +
  scale_color_npg(alpha = 0.7)

# Mitigate effects of cell cycle heterogeneity by regressing out canonical G2/M- and S-Phase
# genes
Rep1_H <- ScaleData(object = Rep1_H,
                    vars.to.regress = c("S.Score",
                                        "G2M.Score"),
                    features = rownames(x = Rep1_H))

# Re-do dimensionality reduction and clustering
Rep1_H_v1 <- Rep1_H %>%
  RunPCA(pc.genes = Rep1_H_v1@var.genes, npcs = 20) %>%
  RunUMAP(reduction = "pca", dims = 1:20) %>%
  FindNeighbors(reduction = "pca", dims = 1:20) %>%
  FindClusters(resolution = 0.2)  ##Note change in resolution

# Visualize the data
DimPlot(Rep1_H_v1,
        reduction = "umap",
        pt.size = 1,
        label = F,
        group.by = "Phase") +
  coord_fixed() +
  ggtitle("Reps by Source") +
  scale_color_npg(alpha = 0.7)

# Run Harmony
Rep1_H_v2 <- Rep1_H %>%
  RunPCA(pc.genes = Rep1_H_v2@var.genes, npcs = 20) %>%
  RunHarmony(group.by.vars = "src", plot_convergence = T) %>%
  RunUMAP(reduction = "harmony", dims = 1:20) %>%
  FindNeighbors(reduction = "harmony", dims = 1:20) %>%
  FindClusters(resolution = 0.2)  ##Note change in resolution

# Visualize the data
DimPlot(Rep1_H_v2,
        reduction = "umap",
        pt.size = 1,
        label = F,
        group.by = "src") +
  coord_fixed() +
  scale_color_npg(alpha = 0.7)

# Run FastMNN
Rep1_H_v3 <- NormalizeData(Rep1_H) %>%
  FindVariableFeatures(selection.method = "vst")

Rep1_H_v3 <- RunFastMNN(object.list = SplitObject(Rep1_H_v3,
                                                  split.by = "cond"))
Rep1_H_v3 <- RunUMAP(Rep1_H_v3,reduction = "mnn", dims = 1:20)
Rep1_H_v3 <- FindNeighbors(Rep1_H_v3, reduction = "mnn", dims = 1:20)
Rep1_H_v3 <- FindClusters(Rep1_H_v3, resolution = 0.2)

DimPlot(Rep1_H_v3,
        reduction = "umap",
        pt.size = 1,
        label = F,
        group.by = "rep") +
  coord_fixed() +
  scale_color_npg(alpha = 0.7)

save(Rep1_H_v3, file = "Data/Supplement/Rep1_H_v3-FastMNN.RData")
save(Rep1_H, file = "Data/Supplement/Rep1_H.RData")
```

```{r s8_a, dependson="s8"}
library(Seurat)
library(tidyverse)

load("Data/Supplement/Rep1_H_v3-FastMNN.RData")

# By rep
sfig_s4_a1 <- DimPlot(Rep1_H_v3,
                      reduction = "umap",
                      pt.size = 1,
                      label = F,
                      group.by = "rep",
                      shuffle = T) +
  theme(aspect.ratio = 1) +
  scale_color_npg(alpha = 1) +
  ggtitle("FastMNN by Replicate") +
  NoAxes()
sfig_s4_a1

# Save as pdf
if(!dir.exists("Figures/supplement/s4")) {
  dir.create("Figures/supplement/s4")
}
pdf("Figures/supplement/s4/sfig_s4_a1.pdf",
    height = 5,
    width = 5)
sfig_s4_a1
dev.off()

# By cell cycle phase
sfig_s4_a2 <- DimPlot(Rep1_H_v3,
                      reduction = "umap",
                      pt.size = 1,
                      label = F,
                      group.by = "Phase",
                      shuffle = T) +
  theme(aspect.ratio = 1) +
  scale_color_npg(alpha = 1) +
  ggtitle("FastMNN by Phase") +
  NoAxes()
sfig_s4_a2

pdf("Figures/supplement/s4/sfig_s4_a2.pdf",
    height = 5,
    width = 5)
sfig_s4_a2
dev.off()

# By src
# Rename src to match paper
Rep1_H_v3$src[grep("Cellecta_Cx", Rep1_H_v3$src)] <- "Rep1 - Cell Culture"
Rep1_H_v3$src[grep("Cellecta_Lung", Rep1_H_v3$src)] <- "Rep1 - Lung"
Rep1_H_v3$src[grep("Zymo_Cx", Rep1_H_v3$src)] <- "Rep2 - Cell Culture"
Rep1_H_v3$src[grep("Zymo_Lung", Rep1_H_v3$src)] <- "Rep2 - Lung"

sfig_s4_a3 <- DimPlot(Rep1_H_v3,
                      reduction = "umap",
                      pt.size = 1,
                      label = F,
                      group.by = "src",
                      shuffle = T) +
  theme(aspect.ratio = 1) +
  #scale_color_npg(alpha = 1) +
  ggtitle("FastMNN by Type") +
  NoAxes()
sfig_s4_a3

pdf("Figures/supplement/s4/sfig_s4_a3.pdf",
    height = 5,
    width = 6)
sfig_s4_a3
dev.off()
```

### S8B

```{r s8_b, dependson="s8"}
library(sctransform)
library(Seurat)
library(ggsci)
options(future.globals.maxSize=89128960000)
set.seed(108)

load("Data/Supplement/Rep1_H.RData")

Rep1_H_sct <- SplitObject(Rep1_H, split.by = "src")

for (i in 1:length(Rep1_H_sct)) {
  Rep1_H_sct[[i]] <- SCTransform(Rep1_H_sct[[i]], verbose = FALSE)
}

Rep1_H.features <- SelectIntegrationFeatures(object.list = Rep1_H_sct,
                                             nfeatures = 3000)
Rep1_H_sct <- PrepSCTIntegration(object.list = Rep1_H_sct,
                                 anchor.features = Rep1_H.features,
                                 verbose = FALSE)
Rep1_H_sct.anchors <- FindIntegrationAnchors(object.list = Rep1_H_sct,
                                             normalization.method = "SCT",
                                             anchor.features = Rep1_H.features,
                                             verbose = FALSE)
Rep1_H_sct.integrated <- IntegrateData(anchorset = Rep1_H_sct.anchors,
                                       normalization.method = "SCT",
                                       verbose = FALSE)

Rep1_H_sct.integrated <- RunPCA(Rep1_H_sct.integrated, verbose = FALSE)
Rep1_H_sct.integrated <- RunUMAP(Rep1_H_sct.integrated, dims = 1:30)

saveRDS(Rep1_H_sct.integrated,
        file = "Data/Supplement/Rep1_H_sct.integrated.rds")
#Rep1_H_sct.integrated <- readRDS("Data/Supplement/Rep1_H_sct.integrated.rds")

sfig_s4_b1 <- DimPlot(Rep1_H_sct.integrated,
                      reduction = "umap",
                      group.by = "rep",
                      pt.size = 1,
                      label = F,
                      shuffle = T) +
  theme(aspect.ratio = 1) +
  ggtitle("SCTransform by Replicate") +
  NoAxes()
sfig_s4_b1

pdf("Figures/supplement/s4/sfig_s4_b1.pdf",
    height = 5,
    width = 5)
sfig_s4_b1
dev.off()

sfig_s4_b2 <- DimPlot(Rep1_H_sct.integrated,
                      reduction = "umap",
                      group.by = "Phase",
                      pt.size = 1,
                      label = F,
                      shuffle = T) +
  theme(aspect.ratio = 1) +
  ggtitle("SCTransform by Phase") +
  scale_color_npg(alpha = 1) +
  NoAxes()
sfig_s4_b2

pdf("Figures/supplement/s4/sfig_s4_b2.pdf",
    height = 5,
    width = 5)
sfig_s4_b2
dev.off()

# Rename src to match paper
Rep1_H_sct.integrated$src[grep("Cellecta_Cx",
                               Rep1_H_sct.integrated$src)] <-
  "Rep1 - Cell Culture"
Rep1_H_sct.integrated$src[grep("Cellecta_Lung",
                               Rep1_H_sct.integrated$src)] <-
  "Rep1 - Lung"
Rep1_H_sct.integrated$src[grep("Zymo_Cx",
                               Rep1_H_sct.integrated$src)] <-
  "Rep2 - Cell Culture"
Rep1_H_sct.integrated$src[grep("Zymo_Lung",
                               Rep1_H_sct.integrated$src)] <-
  "Rep2 - Lung"

sfig_s4_b3 <- DimPlot(Rep1_H_sct.integrated,
                      reduction = "umap",
                      group.by = "src",
                      pt.size = 1,
                      label = F,
                      shuffle = T) +
  theme(aspect.ratio = 1) +
  ggtitle("SCTransform by Type") +
  NoAxes()
sfig_s4_b3

pdf("Figures/supplement/s4/sfig_s4_b3.pdf",
    height = 5,
    width = 5)
sfig_s4_b3
dev.off()
```

### Group

```{r s8_abgroup, dependson=c("s8","s8_a","s8_b")}
s4_ab <- (sfig_s4_a1 | sfig_s4_a2 | sfig_s4_a3) /
  (sfig_s4_b1 | sfig_s4_b2 | sfig_s4_b3)

pdf("Figures/supplement/s4/s4_ab.pdf",
    height=8,
    width=12)
s4_ab
dev.off()
```

### S8C

```{r s8_c}
set.seed(108)
library(harmony)
library(Seurat)
library(tidyverse)
load("Data/OS_list_CCR.RData")

OS_list_harmony <- list()
sfig_s4_c <- list()
for (i in 1:length(OS_list)) {
  # Run Harmony
  OS_list_harmony[[paste0(head(OS_list[[i]]@meta.data$model, 1))]] <- 
    OS_list[[i]] %>%
    RunPCA(pc.genes = OS_list_harmony[[i]]@var.genes, npcs = 20) %>%
    RunHarmony(group.by.vars = "src", plot_convergence = T) %>%
    RunUMAP(reduction = "harmony", dims = 1:20) %>%
    FindNeighbors(reduction = "harmony", dims = 1:20) %>%
    FindClusters(resolution = 0.2)
  
  sfig_s4_c[[i]] <- DimPlot(OS_list_harmony[[i]],
                          reduction = "umap",
                          group.by = "src",
                          pt.size = 1,
                          label = F,
                          shuffle = T) +
    theme(aspect.ratio = 1) +
    ggtitle(paste0(head(OS_list_harmony[[i]]@meta.data$model,
                        1),
                   " Harmony")) +
    NoAxes()
}

pdf("Figures/supplement/s4/sfig_s4_c.pdf",
    width = 16,
    height = 4)
sfig_s4_c[[1]] |
  sfig_s4_c[[2]] |
  sfig_s4_c[[3]] |
  sfig_s4_c[[4]]
dev.off()
```

### Group

```{r s8_group, dependson=c("s8","s8_a","s8_b","s8_c")}
s4 <- (sfig_s4_a1 | sfig_s4_a2 | sfig_s4_a3) /
  (sfig_s4_b1 | sfig_s4_b2 | sfig_s4_b3) /
  (sfig_s4_c[[1]] | sfig_s4_c[[2]] | sfig_s4_c[[3]] | sfig_s4_c[[4]])

pdf("Figures/supplement/s4/s4.pdf",
    height=12,
    width=16)
s4
dev.off()
```

```{r clean2, cache=FALSE}
# BatchEffectCorrectionSCTransform - all
options(future.globals.maxSize= 8912896000000)
# Clean up environment
rm(list=setdiff(ls(), "tenx_load_qc"))
```

# S19-28: Patient Data
(old S5 & S6 & S7)
Patient data

```{r s19-28_processing}
library(rrrSingleCellUtils)
library(patchwork)
library(Seurat)
library(ggplot2)
library(tidyverse)
library(stringr)
# library(future)
library(harmony)
library(devtools)
#source("scSeurat.R")
set.seed(888)

if(!dir.exists("Data/Supplement/PatientData")) {
  tar_dir <- "Data/Supplement/PatientData"
}

#### Raw ####
# plan("multisession", workers = 10)
# options(future.globals.maxSize= 10 * 24000 * 1024^2)

# Make a list of sample names
s <- c("BC5", "BC6", "BC10", "BC11", "BC16",
       "BC17", "BC20", "BC21", "BC22")
qc <- c(18000, 25000, 25000, 30000, 70000, 
        40000, 70000, 50000, 50000)
path <- c("Conventional", "Conventional", "Conventional",
          "Conventional", "Conventional", "Chondroblastic",
          "Chondroblastic", "Intraosseous", "Chondroblastic")
type <- c("Primary", "Primary", "Lung Met", "Primary", "Primary",
          "Lung Met", "Primary", "Primary", "Primary")

# Download, file, and extract the files from GEO
if(!dir.exists("Data/Supplement/PatientData/GSE152048")) {
  tar_dir <- "Data/Supplement/PatientData/GSE152048"
  dir.create(tar_dir)
  geo_pre <- "https://ftp.ncbi.nlm.nih.gov/geo/series/GSE152nnn/GSE152048/suppl/GSE152048_"
  for(i in 1:length(s)){
    gse_path <- str_c(geo_pre, s[i], ".matrix.tar.gz")
    tar_file <- str_c(tar_dir, "/", s[i], ".tar.gz ")
    download.file(gse_path, destfile = tar_file, method = "auto")
    untar(tar_file, exdir = tar_dir)
    file.remove(tar_file)
  }
}

# Create a vector that contains normalized Seurat objects for all samples
raw <- c()
for(i in 1:9) {
  x <- tenx_load_qc(str_c("Data/Supplement/PatientData/GSE152048/",
                          s[i], "/"))
  x <- subset(x, subset = nCount_RNA < qc[i] & percent.mt <13)
  x$src <- s[i]
  x$type <- type[i]
  x$path <- path[i]
  x <- x %>%
    NormalizeData() %>%
    FindVariableFeatures() %>%
    ScaleData() %>%
    RunPCA(verbose = F) %>%
    FindNeighbors(dims = 1:20) %>%
    FindClusters(resolution = 0.3) %>%
    RunUMAP(dims = 1:20)
  print(DimPlot(x, pt.size = 1, label = T) +
          coord_fixed() +
          theme(legend.position = "none") +
          ggtitle(str_c(s[i], " basic clustering")))
  raw <- c(raw, x)
}
rm(x)

#### Comb ####

# Merge into one Seurat object and integrate with harmony
comb <- merge(raw[[1]], y = c(raw[[2]], raw[[3]], raw[[4]], raw[[5]], 
                              raw[[6]], raw[[7]], raw[[8]], raw[[9]]),
              # merge.data = T,
              add.cell.ids = s,
              project = "GSE152048")

comb <- comb %>%
  NormalizeData() %>%
  FindVariableFeatures() %>%
  ScaleData() %>%
  RunPCA(verbose = F)

comb <- RunHarmony(comb, group.by.vars = "src")
comb <- RunUMAP(comb, reduction = "harmony", dims = 1:30)
comb <- comb %>%
  FindNeighbors(reduction = "harmony", dims = 1:30) %>%
  FindClusters()

# Save a stopping point - merged and harmony aligned
qs::qsave(comb, "Data/Supplement/PatientData/comb.qs")
# Start from this stopping point
comb <- qs::qread("Data/Supplement/PatientData/comb.qs")

DimPlot(comb, reduction = "umap", label = T, repel = T) +
  coord_fixed() +
  ggtitle("GSE152048 composite") +
  theme(legend.position = "none")

DimPlot(comb, reduction = "umap", group.by = "src") +
  coord_fixed() +
  ggtitle("GSE152048 composite") 

DimPlot(comb, reduction = "umap", group.by = "type") +
  coord_fixed() +
  ggtitle("GSE152048 composite") 

#### Primary ####

# Separate out the primary tumor lesions
primary <- subset(comb, subset = type == "Primary")

# Ensure that groups are not just dominated by cell cycle effects
#primary <- kill_cc(primary)
# 
# # Set a post-cc-regression stopping point
# saveRDS(primary, "PatientData/primary-ccReg.rds")
# # Start from this stopping point
# primary <- readRDS("PatientData/primary-ccReg.rds")

DimPlot(primary, reduction = "umap", label = T, repel = T) +
  coord_fixed() +
  ggtitle("GSE152048 composite - primaries") +
  theme(legend.position = "none")

DimPlot(primary, reduction = "umap", group.by = "src") +
  coord_fixed() +
  ggtitle("GSE152048 composite - primaries") 

# Calculate and show module scores from the paper
ms <- list()
ms$Osteoblastic <- c("RUNX2", "COL1A1", "CDH11", "IBSP")
ms$Chondroblastic <- c("SOX9", "ACAN", "PTH1R")
ms$Osteoclast <- c("ACP5", "CTSK", "MMP9")
ms$Myeloid <- c("CD74", "CD14", "FCGR3A")
ms$TCell <- c("CD3E", "IL7R", "CD8A", "CD4", "NKG7")
ms$NKCell <- c("NKG7", "GNLY")
ms$NKTCell <- c("NKG7", "GNLY", "CD3E")
ms$DCCell <- c("CD1C", "FCER1A", "CLEC9A", "CCR7", "CD14", "CD163")
ms$Fibroblast <- c("DCN", "COL1A1")
ms$Pericyte <- c("RGS5", "ACTA2")
ms$MSC <- c("MME", "THY1", "CXCL12", "SFRP2")
ms$Endothelial <- c("PECAM1", "VWF")
ms$Myoblast <- c("MYL1", "MYLPF")
ms$BCell <- c("MS4A1", "CD19", "JCHAIN")

mod_names <- names(ms)
#Added min.cutoff to isolate for cells most module like
for(i in 1:length(mod_names)) {
  primary <- AddModuleScore(primary, ms[i], name = names(ms[i]))
  p <- FeaturePlot(primary, features = str_c(names(ms[i]), "1"), min.cutoff = 2, pt.size = 1,
                   order = T, cols = c("lightgoldenrod", "darkred"))
  print(p)
}

#look for proliferation
primary<-kill_cc(primary)
DimPlot(primary, reduction = "umap", label = T, repel = T) +
  coord_fixed() +
  ggtitle("G1/S/G2M cell cycle analysis - primary tumor") +
  theme(legend.position = "none")


#Select Clusters which are tumors. Notice how one group is G2M only. 
#Maybe possible that this group is tumor and due to 0inflation there are transcripts which are not active. 
#Could be cycling back to other tumor states
Idents(primary) <- primary$seurat_clusters
DimPlot(primary, reduction = "umap", label = T, repel = T) +
coord_fixed() +
  ggtitle("GSE152048 composite clusters - primaries Sanj") +
  theme(legend.position = "none")
saveRDS(primary,'Data/Supplement/PatientData/primary.rds')
primary<-readRDS('Data/Supplement/PatientData/primary.rds')

#### Tumor ####

#Re-cluster after subsetting to tumor cells
tumor<- subset(primary, idents=c(2,3,27,16,15,6,28,5,20,29,0,24,14,21,33,35,13,26,7,9,23))
tumor<-tumor%>%
  NormalizeData()%>%
  ScaleData()%>%
  FindVariableFeatures()
tumor<-tumor%>%
  RunPCA(features=VariableFeatures(tumor))%>%
  RunUMAP(dims=1:20)%>%
  FindNeighbors()%>%
  FindClusters(resolution=0.3)

#Plot results
DimPlot(tumor, pt.size=1, reduction="umap", label=T, group.by="src")+
  coord_fixed()+
  ggtitle("Reclustering of Tumor Cells")
#Plot results
DimPlot(tumor, pt.size=1, reduction="umap", label=T)+
  coord_fixed()+
  ggtitle("Reclustering of Tumor Cells")
#kill_cc adds metadata but interferes with analyses
tumor<-kill_cc(tumor)

Idents(tumor) <- tumor$seurat_clusters
DimPlot(tumor, pt.size=1, reduction="umap", label=T)+
  coord_fixed()+
  ggtitle("Reclustering of Tumor Cells")

#Added min.cutoff to isolate for cells most module like
for(i in 1:length(mod_names)) {
  tumor <- AddModuleScore(tumor, ms[i], name = names(ms[i]))
  p <- FeaturePlot(tumor, features = str_c(names(ms[i]), "1"), pt.size = 1,
                   order = T, cols = c("lightgoldenrod", "darkred"))
  print(p)
}
saveRDS(tumor,'Data/Supplement/PatientData/tumor.rds')
tumor<-readRDS('Data/Supplement/PatientData/tumor.rds')

#### Tumor_2 ####

#Re-Subset data further because 7 and 8 look like fibroblast and periocytes modules scored high
tumor_2<-subset(tumor, idents=c(1,2,3,4,5,6,9,10,11,12,13,14))

#Reculuster new tumor
tumor_2<-tumor_2%>%
  NormalizeData()%>%
  ScaleData()%>%
  FindVariableFeatures()
tumor_2<-tumor_2%>%
  RunPCA(features=VariableFeatures(tumor_2))%>%
  RunUMAP(dims=1:20)%>%
  FindNeighbors()%>%
  FindClusters(resolution=0.3)

#Plot results
DimPlot(tumor_2, pt.size=.1, reduction="umap", label=T, group.by="src")+
  coord_fixed()+
  ggtitle("Reclustering of tumor_2 Cells")
#Plot results
DimPlot(tumor_2, pt.size=.1, reduction="umap", label=T)+
  coord_fixed()+
  ggtitle("Reclustering of tumor_2 Cells")
DimPlot(tumor_2, pt.size=.1, reduction="umap", label=T,group.by="seurat_clusters")+
  coord_fixed()+
  ggtitle("Reclustering of tumor_2 Cells")
#kill_cc adds metadata but interferes with analyses
tumor_2<-kill_cc(tumor_2)
Idents(tumor_2)<-tumor$seurat_clusters
for(i in 1:length(mod_names)) {
  tumor_2 <- AddModuleScore(tumor_2, ms[i], name = names(ms[i]))
  p <- FeaturePlot(tumor_2, features = str_c(names(ms[i]), "1"), pt.size = 1,
                   order = T, cols = c("lightgoldenrod", "darkred"))
  print(p)
}
saveRDS(tumor_2,'Data/Supplement/PatientData/tumor_2.rds')
tumor_2<-readRDS('Data/Supplement/PatientData/tumor_2.rds')

#### Tumor_comb ####

#Now that sure of tumor cells. Take the tumor subsets from "tumor" and combine across data with harmony
tumor_comb<-subset(tumor, idents=c(1,2,3,4,5,6,9,10,11,12,13,14))
tumor_comb <- tumor_comb %>%
  NormalizeData() %>%
  FindVariableFeatures() %>%
  ScaleData() %>%
  RunPCA(verbose = F)

tumor_comb <- RunHarmony(tumor_comb, group.by.vars = "src")
tumor_comb <- RunUMAP(tumor_comb, reduction = "harmony", dims = 1:30)
tumor_comb <- tumor_comb %>%
  FindNeighbors(reduction = "harmony", dims = 1:30) %>%
  FindClusters(resolution=0.3)

# Save a stopping point - merged and harmony aligned
qs::qsave(tumor_comb, "Data/Supplement/PatientData/tumor_comb.qs")
# Start from this stopping point
tumor_comb <- qs::qread("Data/Supplement/PatientData/tumor_comb.qs")

DimPlot(tumor_comb, reduction = "umap", label = T, repel = T) +
  coord_fixed() +
  ggtitle("Tumor Clusters post-integration") +
  theme(legend.position = "none")

DimPlot(tumor_comb, reduction = "umap", group.by = "src") +
  coord_fixed() +
  ggtitle("Tumor by sample post-integration") 

tumor_comb<-kill_cc(tumor_comb)

#Run cell type module as before
Idents(tumor_comb)<-tumor_comb$seurat_clusters
for(i in 1:length(mod_names)) {
  tumor_comb <- AddModuleScore(tumor_comb, ms[i], name = names(ms[i]))
  p <- FeaturePlot(tumor_comb, features = str_c(names(ms[i]), "1"), pt.size = 1,
                   order = T, cols = c("lightgoldenrod", "darkred"))
  print(p)
}

#
DimPlot(tumor_comb, reduction = "umap", group.by = "path") +
  coord_fixed() +
  ggtitle("Tumor by pathology type") 

DimPlot(tumor_comb, reduction = "umap", group.by = "type") +
  coord_fixed() +
  ggtitle("Primary or Secondary Tumor")

#### Subset tumor_comb ####

###Get tumor_comb from Ryan, run analysis on individual samples
(table(Idents(tumor_comb)))

#Subset to each sample

s <- c("BC5", "BC6", "BC11", "BC16",
       "BC20", "BC21", "BC22")
primarytumor <- c()
for(i in 1:7) {
  x <- subset(tumor_comb, subset = src == s[i])
  x <- x %>%
    NormalizeData() %>%
    FindVariableFeatures() %>%
    ScaleData() %>%
    RunPCA(verbose = F) %>%
    FindNeighbors(dims = 1:20) %>%
    FindClusters(resolution = 0.1) %>%
    RunUMAP(dims = 1:20)
  print(DimPlot(x, pt.size = 1, label = T) +
          coord_fixed() +
          theme(legend.position = "none") +
          ggtitle(str_c(s[i], " basic clustering")))
  primarytumor <- c(primarytumor, x)
}
rm(x)
names(primarytumor) <- s
################################Set optimum number of clusters using Silhoutte scoring

data <-primarytumor[[7]] 
#data.nC <- nRes(data, 
#                res = seq(from = 0.1, to = 0.9, by = 0.1))
# Optimum resolution for all samples is 0.1, except for BC21 (res = 0.7)
for (i in c(1:5,7)){
  # primarytumor[[i]] <- primarytumor[[i]] %>% FindClusters(resolution = 0.1)
  print(DimPlot(primarytumor[[i]], pt.size = 1, label = T) +
          coord_fixed() +
          theme(legend.position = "none") +
          ggtitle(str_c(s[i], " basic clustering")))
}

for(i in 6){
  primarytumor[[i]] <- primarytumor[[i]] %>% FindClusters(resolution = 0.7)
  print(DimPlot(primarytumor[[i]], pt.size = 1, label = T) +
          coord_fixed() +
          theme(legend.position = "none") +
          ggtitle(str_c(s[i], " basic clustering")))
}

save(primarytumor,
     file = "Data/Supplement/primarytumor_preModuleScoring.RData")
```

## S19-22 Module score FeaturePlots

```{r s19-28_modules, dependson="s19-28_processing"}
# Set up directory
if(!dir.exists("Figures/supplement/s5")) {
  dir.create("Figures/supplement/s5")
}
# For s5
# Apply Glycolysis module score to each of the primary tumor cells
gl <- list()
Glycolysis <- read.table(file = "pathways/Glycolysis.txt",
                         header = FALSE,
                         sep = ",")
Glycolysis <- Glycolysis[, 1]

# for s6
Hypoxia <- read.table(file = "pathways/Hypoxia.txt",
                      header = FALSE,
                      sep = ",")
Hypoxia <- Hypoxia[, 1]

# For S7
EMT <- read.table(file = "pathways/EMT.txt",
                  header = FALSE,
                  sep = ",")
EMT <- EMT[, 1]

# For s8
TNFA <- read.table(file = "pathways/TNFA.txt",
                   header = FALSE,
                   sep = ",")
TNFA <- TNFA[, 1]

gl$Glycolysis <- Glycolysis
gl$Hypoxia <- Hypoxia
gl$EMT <- EMT
gl$TNFA <- TNFA

mod_names <- names(gl)

#### S5 ####
# Glycolysis - Added min.cutoff to isolate for cells most module like
gly <- list()
for (i in 1:length(primarytumor)) {
  primarytumor[[i]] <- AddModuleScore(primarytumor[[i]],
                             gl[1],
                             name = names(gl[1]))
  p <- FeaturePlot(primarytumor[[i]],
                   features = str_c(names(gl[1]),
                                    "1"),
                   min.cutoff = 0.1,
                   pt.size = 1,
                   order = T,
                   cols = c("lightgoldenrod",
                            "darkred")) +
    ggtitle(paste0(s[i])) +
    theme(aspect.ratio = 1)
  
  pdf(file = paste0("Figures/supplement/s5/",
                    names(gl[1]),
                    "_FeaturePlot_",
                    s[i],
                    ".pdf"))
  print(p)
  dev.off()
  
  gly[[i]] <- p
  
}

s5 <- gly[[1]] +
  gly[[2]] +
  gly[[3]] +
  gly[[4]] +
  gly[[5]] +
  gly[[6]] +
  gly[[7]] +
  plot_layout(ncol = 3)

pdf(file = "Figures/supplement/s5/s5.pdf",
    height=13,
    width=15)
s5
dev.off()

#### S6 ####
# Hypoxia - Added min.cutoff to isolate for cells most module like
hypx <- list()
for (i in 1:length(primarytumor)) {
  primarytumor[[i]] <- AddModuleScore(primarytumor[[i]],
                             gl[2],
                             name = names(gl[2]))
  p <- FeaturePlot(primarytumor[[i]],
                   features = str_c(names(gl[2]),
                                    "1"),
                   min.cutoff = 0.1,
                   pt.size = 1,
                   order = T,
                   cols = c("lightgoldenrod",
                            "darkred")) +
    ggtitle(paste0(s[i])) +
    theme(aspect.ratio = 1)

  if (!dir.exists("Figures/supplement/s6")) {
    dir.create("Figures/supplement/s6")
  }
  pdf(file = paste0("Figures/supplement/s6/",
                    names(gl[2]),
                    "_FeaturePlot_",
                    s[i],
                    ".pdf"))
  print(p)
  dev.off()
  
  hypx[[i]] <- p
}
s6 <- hypx[[1]] +
  hypx[[2]] +
  hypx[[3]] +
  hypx[[4]] +
  hypx[[5]] +
  hypx[[6]] +
  hypx[[7]] +
  plot_layout(ncol = 3)

pdf(file = "Figures/supplement/s6/s6.pdf",
    height=13,
    width=15)
s6
dev.off()

#### S7 ####
# EMT - Added min.cutoff to isolate for cells most module like
emt <- list()
for (i in 1:length(primarytumor)) {
  primarytumor[[i]] <- AddModuleScore(primarytumor[[i]],
                             gl[3],
                             name = names(gl[3]))
  p <- FeaturePlot(primarytumor[[i]],
                   features = str_c(names(gl[3]),
                                    "1"),
                   min.cutoff = 0.1,
                   pt.size = 1,
                   order = T,
                   cols = c("lightgoldenrod",
                            "darkred")) +
    ggtitle(paste0(s[i])) +
    theme(aspect.ratio = 1)
  
    if (!dir.exists("Figures/supplement/s7")) {
    dir.create("Figures/supplement/s7")
  }
  pdf(file = paste0("Figures/supplement/s7/",
                    names(gl[3]),
                    "_FeaturePlot_",
                    s[i],
                    ".pdf"))
  print(p)
  dev.off()
  
  emt[[i]] <- p
}

s7 <- emt[[1]] +
  emt[[2]] +
  emt[[3]] +
  emt[[4]] +
  emt[[5]] +
  emt[[6]] +
  emt[[7]] +
  plot_layout(ncol = 3)

pdf(file = "Figures/supplement/s7/s7.pdf",
    height=13,
    width=15)
s7
dev.off()

#### S8 ####
# TNFA - Added min.cutoff to isolate for cells most module like
tnf <- list()
for (i in 1:length(primarytumor)) {
  primarytumor[[i]] <- AddModuleScore(primarytumor[[i]],
                             gl[4],
                             name = names(gl[4]))
  p <- FeaturePlot(primarytumor[[i]],
                   features = str_c(names(gl[4]),
                                    "1"),
                   min.cutoff = 0.1,
                   pt.size = 1,
                   order = T,
                   cols = c("lightgoldenrod",
                            "darkred")) +
    ggtitle(paste0(s[i])) +
    theme(aspect.ratio = 1)
  
    if (!dir.exists("Figures/supplement/s8")) {
    dir.create("Figures/supplement/s8")
  }
  pdf(file = paste0("Figures/supplement/s8/",
                    names(gl[4]),
                    "_FeaturePlot_",
                    s[i],
                    ".pdf"))
  print(p)
  dev.off()
  
  tnf[[i]] <- p
}

s8 <- tnf[[1]] +
  tnf[[2]] +
  tnf[[3]] +
  tnf[[4]] +
  tnf[[5]] +
  tnf[[6]] +
  tnf[[7]] +
  plot_layout(ncol = 3)

pdf(file = "Figures/supplement/s8/s8.pdf",
    height=13,
    width=15)
s8
dev.off()

save(primarytumor, file = "Data/Supplement/PatientData/primarytumor.RData")
```

## S22-28 Module score ScatterPlots

```{r s19-28_scatter, dependson=c("s19-28_processing","s19-28_modules")}
#print(hue_pal()(7))
colors <- c("#F8766D",
            "#C49A00",
            "#53B400",
            "#00C094",
            "#00B6EB",
            "#A58AFF",
            "#FB61D7")

#### Glycolysis vs all others ####
combo <- list()
for (i in 1:3) {
  for (j in 1:length(primarytumor)) {
    p <- FeatureScatter(primarytumor[[j]],
                        feature1 = str_c(names(gl[1]),
                                         "1"),
                        
                        feature2 = str_c(names(gl[i+1]),
                                         "1"),
                        group.by = "src",
                        cols = colors[[j]],
                        pt.size = 0.5) +
      theme(aspect.ratio = 1)
    print(p)
    
    # Set up directory
    if(!dir.exists(paste0("Figures/supplement/s5-",
                          i+5))) {
      dir.create(paste0("Figures/supplement/s5-",
                        i+5))
    }
    
    pdf(file = paste0("Figures/supplement/s5-",
                      i+5,
                      "/",
                      names(gl[1]),
                      names(gl[i+1]),
                      "_ScatterPlot_",
                      s[j],
                      ".pdf"),
        height = 5,
        width = 5)
    print(p)
    dev.off()
    
    combo[[j]] <- p
  }
  sfig <- combo[[1]] +
    combo[[2]] +
    combo[[3]] +
    combo[[4]] +
    combo[[5]] +
    combo[[6]] +
    combo[[7]] +
    plot_layout(ncol = 3)

    pdf(file = paste0("Figures/supplement/s5-",
                      i+5,
                      "/s5-",
                      i+5,
                      ".pdf"),
        height=13,
        width=15)
    print(sfig)
    dev.off()
}

#### Hypoxia vs Remaining ####
combo <- list()
for (i in 1:2) {
  for (j in 1:length(primarytumor)) {
    p <- FeatureScatter(primarytumor[[j]],
                        feature1 = str_c(names(gl[2]),
                                         "1"),
                        
                        feature2 = str_c(names(gl[i+2]),
                                         "1"),
                        group.by = "src",
                        cols = colors[[j]],
                        pt.size = 0.5) +
      theme(aspect.ratio = 1)
    print(p)
    
      # Set up directory
  if(!dir.exists(paste0("Figures/supplement/s6-",
                      i+6))) {
    dir.create(paste0("Figures/supplement/s6-",
                      i+6))
               }
    pdf(file = paste0("Figures/supplement/s6-",
                      i+6,
                      "/",
                      names(gl[2]),
                      names(gl[i+2]),
                      "_ScatterPlot_",
                      s[j],
                      ".pdf"),
        height = 5,
        width = 5)
    print(p)
    dev.off()
    
    combo[[j]] <- p
  }
    sfig <- combo[[1]] +
      combo[[2]] +
      combo[[3]] +
      combo[[4]] +
      combo[[5]] +
      combo[[6]] +
      combo[[7]] +
      plot_layout(ncol = 3)
    
    pdf(file = paste0("Figures/supplement/s6-",
                      i+6,
                      "/s6-",
                      i+6,
                      ".pdf"),
        height=13,
        width=15)
    print(sfig)
    dev.off()
}

#### Last Two vs Each other ####
combo <- list()

  for (i in 1:length(primarytumor)) {
    p <- FeatureScatter(primarytumor[[i]],
                        feature1 = str_c(names(gl[3]),
                                         "1"),
                        
                        feature2 = str_c(names(gl[4]),
                                         "1"),
                        group.by = "src",
                        cols = colors[[i]],
                        pt.size = 0.5) +
      theme(aspect.ratio = 1)
    print(p)
    
    # Set up directory
    if(!dir.exists("Figures/supplement/s7-8")) {
      dir.create("Figures/supplement/s7-8")
    }
    pdf(file = paste0("Figures/supplement/s7-8/",
                      names(gl[3]),
                      names(gl[4]),
                      "_ScatterPlot_",
                      s[i],
                      ".pdf"),
        height = 5,
        width = 5)
    print(p)
    dev.off()
    
    combo[[i]] <- p
  }
    sfig <- combo[[1]] +
      combo[[2]] +
      combo[[3]] +
      combo[[4]] +
      combo[[5]] +
      combo[[6]] +
      combo[[7]] +
      plot_layout(ncol = 3)
    
    pdf(file = paste0("Figures/supplement/s7-8/s7-8.pdf"),
        height=13,
        width=15)
    print(sfig)
    dev.off()

```

# S29 & S30: Imaging
(old S9 & S10 - Imaging)

# S32
(old S11)

## Processing & S32B (cell culture)

```{r s32_1}
#source("scSeurat.R")
#loading libraries
library(Seurat)
library(future)
library(ggplot2)
# library(fgsea)
library(msigdbr)
library(clusterProfiler)
library(SingleR)
library(dplyr) # %<%
library(pheatmap)
library(RColorBrewer)
library(viridis) # inferno color palette
library(rrrSingleCellUtils)
set.seed(100)

#Cellecta and Zymo, Cell culture 
#Cellecta
cellecta.cx.raw <- tenx_load_qc(
  "/gpfs0/home2/gdrobertslab/lab/Counts/S0015-cellecta/outs/filtered_feature_bc_matrix/",
  species_pattern = "^hg19-")
cellecta.cx.raw <- subset(cellecta.cx.raw,
                          subset = nFeature_RNA > 2500 &
                            nCount_RNA < 40000 &
                            percent.mt < 14)
cellecta.cx.raw$src <- "Replicate_02"
cellecta.cx.raw$cond <- "Culture"

#Zymo
zymo.cx.raw <- tenx_load_qc(
  path_10x = "/gpfs0/home/gdrobertslab/lab/Counts/S0016xS0027/outs/filtered_feature_bc_matrix/",
  species_pattern = "^hg19-")
zymo.cx.raw <- subset(zymo.cx.raw,
                      subset = nFeature_RNA > 3500 &
                        nCount_RNA < 100000 &
                        percent.mt < 15)
zymo.cx.raw$src <- "Replicate_01"
zymo.cx.raw$cond <- "Culture"

# Create table of lineage tags for cells in cellecta replicate
if(!file.exists("Data/S0015-cellecta.ltbc")){
  gen_cellecta_bc_data(
    file = "/gpfs0/home/gdrobertslab/lab/Counts/S0015-cellecta/outs/possorted_genome_bam.bam",
    verbose = TRUE,
    output = "Data/S0015-cellecta.ltbc",
    samtools_module = "SAMtools")
  ltbc <- read.table("Data/S0015-cellecta.ltbc",
                     header = T,
                     sep = "\t")
  if(substr(ltbc[[1, 1]], 1, 5)=="CB:Z:") {
    ltbc$cid <- substr(ltbc$cid, 6, 22)
    write.table(ltbc, "Data/S0015-cellecta.ltbc", sep = "\t")
  }
}

# Add lineage tags to metadata of objects
cellecta.cx.raw  <- process_ltbc(cellecta.cx.raw ,
                                 cid_lt = read.table("Data/S0015-cellecta.ltbc",
                                                     header = T,
                                                     sep = "\t"),
                                 histogram = F,
                                 relative = F,
                                 ret_list = F)

zymo.cx.raw <- process_ltbc(zymo.cx.raw,
                            cid_lt = read.table("Data/S0016xS0027.ltbc",
                                                header = T,
                                                sep = "\t"),
                            histogram = F,
                            relative = F,
                            ret_list = F)

# Cellecta
n_uniq_1 <-
  (table(cellecta.cx.raw$lt) == 1) %>%
  sum()

# cellecta.cx.raw <- process_ltbc(cellecta.cx.raw,
#                                 cid_lt = read.table("Data/S0015-cellecta.ltbc",
#                                                     header = T,
#                                                     sep = "\t"),
#                                 histogram = F,
#                                 title = "Cx - Cellecta, Top 40 Clones",
#                                 ymax = 10,
#                                 col_fill = "#E64B35FF",
#                                 relative = T,
#                                 ret_list = F)

# Zymo
n_uniq_2 <-
  (table(zymo.cx.raw$lt) == 1) %>%
  sum()

# zymo.cx.raw <- process_ltbc(zymo.cx.raw,
#                             cid_lt = read.table("Data/S0016xS0027.ltbc",
#                                                 header = T,
#                                                 sep = "\t"),
#                             histogram = F,
#                             title = "Cx - Zymo, Top 40 Clones",
#                             ymax = 2.5,
#                             col.fill = "#00A087FF",
#                             relative = T,
#                             ret_list = F)

# Merge into a single Seurat object
Reps <- merge(cellecta.cx.raw,
              y = zymo.cx.raw,
              add.cell.ids = c("Cellecta_Cx",
                               "Zymo_Cx"),
              project = "Heterogeneity")

# Process and cluster
Reps <- NormalizeData(Reps) %>%
  FindVariableFeatures(selection.method = "vst") %>%
  ScaleData() %>%
  RunPCA(pc.genes = Reps@var.genes, npcs = 20) %>%
  RunUMAP(reduction = "pca", dims = 1:20) %>%
  FindNeighbors(reduction = "pca", dims = 1:20) %>%
  FindClusters(resolution = 0.2)

s.genes <- cc.genes$s.genes
g2m.genes <- cc.genes$g2m.genes
Reps <- CellCycleScoring(object = Reps,
                         s.features = s.genes,
                         g2m.features = g2m.genes,
                         set.ident = TRUE)

# Attempt to regress out the effects of cell cycle on these tumor cells
Reps <- CellCycleScoring(object = Reps,
                         s.features = s.genes,
                         g2m.features = g2m.genes,
                         set.ident = TRUE)

Reps <- ScaleData(object = Reps,
                  vars.to.regress = c("S.Score",
                                      "G2M.Score"),
                  features = rownames(x = Reps))

Reps <- RunPCA(Reps, pc.genes = Reps@var.genes, npcs = 20) %>%
  RunUMAP(reduction = "pca", dims = 1:20) %>%
  FindNeighbors(reduction = "pca", dims = 1:20) %>%
  FindClusters(resolution = 0.3)

# Save
save(Reps, file = "Data/Supplement/Reps_s11_1_ccr.RData")

# Set up directory
    if(!dir.exists("Figures/supplement/s11")) {
      dir.create("Figures/supplement/s11")
    }
# Plot the data 
sfig_s11_b1 <- DimPlot(Reps,
                       reduction = "umap",
                       group.by = "src",
                       pt.size = 1,
                       label = F,
                       shuffle = T) + 
  theme(aspect.ratio = 1) + 
  ggtitle("Reps by Source") + 
  scale_color_npg(alpha = 1) +
  ggtitle("OS17 in Cell Culture (replicates)")
sfig_s11_b1

pdf("Figures/supplement/s11/sfig_s11_b1.pdf",
    width = 6,
    height = 6)
sfig_s11_b1
dev.off()
```

## Processing & S32B (lung)

```{r s32_2}
#source("scSeurat.R")
library(dplyr)
library(ggplot2)
library(harmony)
library(cowplot)
library(Seurat)
library(rrrSingleCellUtils)
set.seed(100)

# Cellecta
cellecta.lung.raw <- tenx_load_qc("/gpfs0/home2/gdrobertslab/lab/Counts/S0023-cellecta-lung/outs/filtered_feature_bc_matrix/",
                                species_pattern = "^hg19-")
cellecta.lung.raw <- subset(cellecta.lung.raw,
                            subset = percent.mt < 20 &
                              percent.mt > 2)
cellecta.lung.raw$src <- "Replicate_02"
cellecta.lung.raw$cond <- "Lung"

#Zymo
zymo.lung.raw<- tenx_load_qc("/gpfs0/home2/gdrobertslab/lab/Counts/S0024xS0029/outs/filtered_feature_bc_matrix/",
                           species_pattern = "^hg19-")
zymo.lung.raw <- subset(zymo.lung.raw,
                        subset = nFeature_RNA > 1250 &
                          nCount_RNA < 60000 &
                          percent.mt < 25)
zymo.lung.raw$src <- "Replicate_01"
zymo.lung.raw$cond <- "Lung"

# Add lineage tracing tags to the Seurat objects
# Create table of lineage tags for cells in cellecta replicate
if(!file.exists("Data/S0023-cellecta-lung.ltbc")){
  gen_cellecta_bc_data(
    file = "/gpfs0/home/gdrobertslab/lab/Counts/S0023-cellecta-lung/outs/possorted_genome_bam.bam",
    verbose = TRUE,
    output = "Data/S0023-cellecta-lung.ltbc",
    samtools_module = "SAMtools")
  ltbc <- read.table("Data/S0023-cellecta-lung.ltbc",
                     header = T,
                     sep = "\t")
  if(substr(ltbc[[1, 1]], 1, 5)=="CB:Z:") {
    ltbc$cid <- substr(ltbc$cid, 6, 22)
    write.table(ltbc, "Data/S0023-cellecta-lung.ltbc", sep = "\t")
  }
}

# Add lineage tags to metadata of objects
cellecta.lung.raw  <- process_ltbc(cellecta.lung.raw ,
                                   cid_lt = read.table("Data/S0023-cellecta-lung.ltbc",
                                                       header = T,
                                                       sep = "\t"),
                                   histogram = F,
                                   relative = F,
                                   ret_list = F)

zymo.lung.raw <- process_ltbc(zymo.lung.raw,
                              cid_lt = read.table("Data/S0024xS0029.ltbc",
                                                  header = T,
                                                  sep = "\t"),
                              histogram = F,
                              relative = F,
                              ret_list = F)

# Cellecta
n_uniq_3 <-
  (table(cellecta.lung.raw$lt) == 1) %>%
  sum()

# Zymo
n_uniq_4 <-
  (table(zymo.lung.raw$lt) == 1) %>%
  sum()

# Merge into a single Seurat object
Reps2 <- merge(cellecta.lung.raw,
               y = zymo.lung.raw,
               add.cell.ids = c("Cellecta_lung",
                                "Zymo_lung"),
               project = "Heterogeneity")

# Process and cluster
Reps2 <- NormalizeData(Reps2) %>%
  FindVariableFeatures(selection.method = "vst") %>%
  ScaleData() %>%
  RunPCA(pc.genes = Reps2@var.genes, npcs = 20) %>%
  RunUMAP(reduction = "pca", dims = 1:20) %>%
  FindNeighbors(reduction = "pca", dims = 1:20) %>%
  FindClusters(resolution = 0.2) 

s.genes <- cc.genes$s.genes
g2m.genes <- cc.genes$g2m.genes
Reps2 <- CellCycleScoring(object = Reps2,
                          s.features = s.genes,
                          g2m.features = g2m.genes,
                          set.ident = TRUE)

# Attempt to regress out the effects of cell cycle on these tumor cells
Reps2 <- CellCycleScoring(object = Reps2,
                          s.features = s.genes,
                          g2m.features = g2m.genes,
                          set.ident = TRUE)

Reps2 <- ScaleData(object = Reps2,
                   vars.to.regress = c("S.Score",
                                       "G2M.Score"),
                   features = rownames(x = Reps2))

Reps2 <- RunPCA(Reps2, pc.genes = Reps2@var.genes, npcs = 20) %>%
  RunUMAP(reduction = "pca", dims = 1:20) %>%
  FindNeighbors(reduction = "pca", dims = 1:20) %>%
  FindClusters(resolution = 0.3)

save(Reps2, file = "Data/Supplement/Reps_s11_2_lung_ccr.RData")

# Plot the data 
sfig_s11_b2 <- DimPlot(Reps2,
                       reduction = "umap",
                       group.by = "src",
                       pt.size = 1,
                       label = F,
                       shuffle = T) + 
  theme(aspect.ratio = 1) + 
  ggtitle("Reps by Source") + 
  scale_color_npg(alpha = 1) +
  ggtitle("OS17 in Orthotopic Lung (replicates)") 
sfig_s11_b2

pdf("Figures/supplement/s11/sfig_s11_b2.pdf",
    width = 6,
    height = 6)
sfig_s11_b2
dev.off()
```

## S32A and Panel Grouping

```{r s32_correctplots, dependson=c("s32_1","s32_2")}
library(rrrSingleCellUtils)
# Histogram of linage tags
ltbc_hist <- function(df, fill, n_uniq, title, subtitle) {
  df %>% 
    head(n = 40) %>%
    ggplot(aes(x = stats::reorder(Var1, -freq),
                   y = freq)) +
    geom_bar(fill = fill, stat = "identity") +
    theme_bw() +
    theme(plot.title = element_text(size = 8),
          axis.title = element_text(size = 6),
          axis.text = element_text(size = 6),
          axis.text.x = element_blank(),
          axis.ticks.x = element_blank(),
          panel.grid.major = element_blank(),
          panel.grid.minor = element_blank()) +
    labs(x = "",
         y = "Percentage of Cells",
         title = title,
         subtitle = subtitle) +
    ylim(0, 10) +
    annotate("text",
             label = paste(n_uniq,
                           "out of every 1000 clones are unique"),
             x = -Inf,
             y = Inf,
             hjust = -0.4,
             vjust = 3,
             size = 2)
}

sfig_s11_a1 <- 
  process_ltbc(cellecta.cx.raw,
               cid_lt = read.table("Data/S0015-cellecta.ltbc",
                                   header = T,
                                   sep = "\t"),
               histogram = FALSE,
               relative = TRUE,
               ret_list = TRUE) %>%
  ltbc_hist(fill = "#E64B35FF",
            n_uniq_1,
            "Replicate #2 Lineage Tag Enrichment Analysis",
            subtitle = "Culture")

pdf("Figures/supplement/s11/sfig_s11_a1.pdf",
    width = 8,
    height = 2)
sfig_s11_a1 
dev.off()

sfig_s11_a2 <-  process_ltbc(cellecta.lung.raw,
                             cid_lt = read.table("Data/S0023-cellecta-lung.ltbc",
                                                 header = T,
                                                 sep = "\t"),
                             histogram = FALSE,
                             relative = TRUE,
                             ret_list = TRUE) %>%
  ltbc_hist(fill = "#E64B35FF",
            n_uniq_3,
            "",
            subtitle = "Lung")

pdf("Figures/supplement/s11/sfig_s11_a2.pdf",
    width = 8,
    height = 2)
sfig_s11_a2
dev.off()

#Patchwork 
layout <- "
AAAACCDD
BBBBCCDD
"
s11 <- sfig_s11_a1 +
  sfig_s11_a2 +
  sfig_s11_b1 +
  sfig_s11_b2 +
  plot_layout(design = layout)
s11
pdf("Figures/supplement/s11/s11.pdf",
    width = 25,
    height = 5)
s11
dev.off()

#v2
layout <- "
AAAA
BBBB
CCDD
CCDD
"
s11 <- sfig_s11_a1 +
  sfig_s11_a2 +
  sfig_s11_b1 +
  sfig_s11_b2 +
  plot_layout(design = layout)
s11
pdf("Figures/supplement/s11/s11v2.pdf",
    width = 12,
    height = 9)
s11
dev.off()
```

```{r session}
sessionInfo()
```
