---
title: "OSHetero2021 Loading and Processing"
author: "Sanjana Rajan and Emily Franz"
date: "`r format(Sys.time(), '%m/%d/%Y')`"
output: 
  html_document:
    toc: true
    toc_float: true
    toc_depth: 5
    number_sections: false
    code_folding: hide
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
  tidy = TRUE,
  echo = TRUE,
  cache = TRUE,
  collapse = TRUE,
  tidy.opts = list(width.cutoff = 95),
  message = FALSE,
  warning = FALSE,
  cache.lazy = FALSE,
  fig.show='hide'
)
```

```{r Fig1_lib, cache=FALSE}
source("Downstream.v2.R")

# loading libraries
library(Seurat)
library(rrrSingleCellUtils)
library(ggplot2)
library(msigdbr)
library(dplyr)
```

# Load Seurat objects 

Load Seurat objects here so only have to complete once (commented out elsewhere). 

## Load Osteoblasts

```{r load_ob}
# Create Seurat objects and perform initial QC.  Label original source.
# osteoblasts
osteoblasts <- tenx_load_qc(
  path_10x = "/gpfs0/home/gdrobertslab/lab/Counts/S0031/outs/filtered_feature_bc_matrix/"
  )

osteoblasts <- subset(osteoblasts,
  subset = nCount_RNA < 20000 &
  percent.mt < 10
  )

osteoblasts$src <- "Osteoblasts"
osteoblasts$model <- "Osteoblasts"
```

## Load OS17

```{r load_os17}
# OS17
## Culture
os17_cx_raw <- tenx_load_qc(
  path_10x = "/gpfs0/home/gdrobertslab/lab/Counts/S0016xS0027/outs/filtered_feature_bc_matrix/",
  species_pattern = "^hg19-"
)

os17_cx_raw <- subset(os17_cx_raw,
  subset = nFeature_RNA > 3500 &
  nCount_RNA < 100000 &
  percent.mt < 15
)

os17_cx_raw$src <- "Culture"
os17_cx_raw$model <- "OS-17"

## Tibia
os17_tib_raw <- tenx_load_qc(
  path_10x = "/gpfs0/home/gdrobertslab/lab/Counts/S0018xS0028/outs/filtered_feature_bc_matrix/",
  species_pattern = "^hg19-"
)

os17_tib_raw <- subset(os17_tib_raw,
  subset = nFeature_RNA > 3000 &
  nCount_RNA < 60000 &
  percent.mt < 18
)

os17_tib_raw$src <- "Tibia"
os17_tib_raw$model <- "OS-17"

## Lung
## double check usage throughout rest of code - Emily (fixed issue where coming from wrong file)
os17_lung_raw <- tenx_load_qc(
  path_10x = "/gpfs0/home2/gdrobertslab/lab/Counts/S0024xS0029/outs/filtered_feature_bc_matrix",
  species_pattern = "^hg19-")

os17_lung_raw <- subset(os17_lung_raw,
  subset = nFeature_RNA > 1250 &
  nCount_RNA < 60000 &
  percent.mt < 25)

os17_lung_raw$src <- "Lung"
os17_lung_raw$model <- "OS-17"
```

### Process OS17 culture for Figure 1

```{r Fig1a1_process, dependson='load_os17', fig.height=8, fig.width=8}
os17_cx <- NormalizeData(os17_cx_raw) %>%
  FindVariableFeatures(selection.method = "vst") %>%
  ScaleData() %>%
  RunPCA(pc.genes = os17_cx@var.genes, npcs = 20) %>%
  RunUMAP(reduction = "pca", dims = 1:20) %>%
  FindNeighbors(reduction = "pca", dims = 1:20) %>%
  FindClusters(resolution = 0.3)

DimPlot(os17_cx,
  reduction = "umap",
  pt.size = 1,
  label = T) +
  coord_fixed() +
  scale_color_npg(alpha = 0.7)

if(!file.exists("Data")){
  dir.create("Data")
}
save(os17_cx, file = "Data/os17_cx_noccr.RData")

# Regress out the effects of cell cycle on these tumor cells
s.genes <- cc.genes$s.genes
g2m.genes <- cc.genes$g2m.genes
os17_cx <- CellCycleScoring(object = os17_cx,
  s.features = s.genes,
  g2m.features = g2m.genes,
  set.ident = TRUE)
os17_cx <- ScaleData(object = os17_cx,
  vars.to.regress = c("S.Score", "G2M.Score"),
  features = VariableFeatures(os17_cx),
  block.size = 10000)

os17_cx <- RunPCA(os17_cx,
  pc.genes = os17_cx@var.genes,
  npcs = 20) %>%
  RunUMAP(reduction = "pca",
    dims = 1:20) %>%
  FindNeighbors(reduction = "pca",
    dims = 1:20) %>%
  FindClusters(resolution = 0.3)

# Find optimal clustering resolution
os17_cx_nc <- nRes(os17_cx, 
  res = seq(from = 0.1, to = 0.2, by = 0.05))
plot <- pSil(os17_cx_nc, 0.15)
plot

# Therefore, we decided to proceed with res = 0.15
os17_cx <- FindClusters(os17_cx, resolution = 0.15)

if(!file.exists("Data")) {
  dir.create("Data")
}

save(os17_cx, file = "Data/os17_cx_CCR.RData")
```

### Add lineage tags to metadata for Figure 5

For lineage analysis in Figure 5.

```{r Fig5_tag, dependson='load_os17'}
# Add lineage tracing tags to the Seurat object metadata
ltbc_list <- c("S0016xS0027",
               "S0018xS0028",
               "S0024xS0029")

for (ltbc_name in ltbc_list){
  if(!file.exists(paste0("Data/", ltbc_name, ".ltbc"))){
    gen_cellecta_bc_data(
      file = paste0("/gpfs0/home/gdrobertslab/lab/Counts/",
                    ltbc_name,
                    "/outs/possorted_genome_bam.bam"),
      verbose = TRUE,
      output = paste0("Data/", ltbc_name, ".ltbc"),
      samtools_module = "SAMtools")
    ltbc <- read.table(paste0("Data/", ltbc_name, ".ltbc"), header = T, sep = "\t")
    if(substr(ltbc[[1, 1]], 1, 5)=="CB:Z:") {
      ltbc$cid <- substr(ltbc$cid, 6, 22)
      write.table(ltbc, paste0("Data/", ltbc_name, ".ltbc"), sep = "\t")
    }
  }
}

os17_cx_raw <- process_ltbc(os17_cx_raw,
                            cid_lt = read.table("Data/S0016xS0027.ltbc",
                                                header = T,
                                                sep = "\t"),
                            histogram = F,
                            relative = T)

os17_tib_raw <- process_ltbc(os17_tib_raw,
                             cid_lt = read.table("Data/S0018xS0028.ltbc",
                                                 header = T,
                                                 sep = "\t"),
                             histogram = F,
                             relative = T)

os17_lung_raw <- process_ltbc(os17_lung_raw,
                              cid_lt = read.table("Data/S0024xS0029.ltbc",
                                                  header = T,
                                                  sep = "\t"),
                              histogram = F,
                              relative = T)
```

### Create OS17 object for lineage analysis in Figure 5

```{r Fig5_process, dependson='Fig5_tag', fig.height=8, fig.width=8}
# Subset all to 2800 cells in each condition
set.seed(108)
os17_cx <- subset(os17_cx_raw,
  cells = sample(Cells(os17_cx_raw), 2800))
os17.tib <- subset(os17_tib_raw,
  cells = sample(Cells(os17_tib_raw), 2800))
os17.lung <- subset(os17_lung_raw,
  cells = sample(Cells(os17_lung_raw), 2800))

# Merge into a single Seurat object
os17 <- merge(os17_cx,
  y = c(os17.tib, os17.lung),
  add.cell.ids = c("Culture", "Tibia", "Lung"),
  project = "LineageTracing")

# Process and cluster
os17 <- NormalizeData(os17) %>%
  FindVariableFeatures(selection.method = "vst") %>%
  ScaleData() %>%
  RunPCA(pc.genes = os17@var.genes, npcs = 20) %>%
  RunUMAP(reduction = "pca", dims = 1:20) %>%
  FindNeighbors(reduction = "pca", dims = 1:20) %>%
  FindClusters(resolution = 0.3)

# CCR Attempt to regress out the effects of cell cycle on these tumor cells
s.genes <- cc.genes$s.genes
g2m.genes <- cc.genes$g2m.genes

os17 <- CellCycleScoring(object = os17,
  s.features = s.genes,
  g2m.features = g2m.genes,
  set.ident = TRUE)

os17 <- ScaleData(object = os17,
  vars.to.regress = c("S.Score", "G2M.Score"),
  features = VariableFeatures(os17),
  block.size = 10000)

os17 <- RunPCA(os17, pc.genes = os17@var.genes, npcs = 20) %>%
  RunUMAP(reduction = "pca", dims = 1:20) %>%
  FindNeighbors(reduction = "pca", dims = 1:20) %>%
  FindClusters(resolution = 0.3)

DimPlot(os17,
    reduction = "umap",
    pt.size = 1,
    label = T) +
  coord_fixed() +
  ggtitle("OS17 Clusters") +
  scale_color_npg(alpha = 0.7)

# Plot the data
set.seed(100)
cell_ids <- sample(colnames(os17))

DimPlot(os17,
    reduction = "umap",
    group.by = "src",
    pt.size = 1,
    label = F,
    order = cell_ids) +
  coord_fixed() +
  ggtitle("OS17 by Source") +
  scale_color_npg()

save(os17, file = "Data/os17.RData")
```

## Load t143b

```{r load_t143b}
# t143b
## Culture
t143b_cx_raw <- tenx_load_qc(
  path_10x = "/gpfs0/home/gdrobertslab/lab/Counts/S0017/outs/filtered_feature_bc_matrix/",
  species_pattern = "^hg19-")

t143b_cx_raw <- subset(t143b_cx_raw,
  subset = nFeature_RNA > 4000 &
  nCount_RNA < 74000 &
  percent.mt < 17)

t143b_cx_raw$src <- "Culture"
t143b_cx_raw$model <- "143B"

## Tibia
t143b_tib_raw <- tenx_load_qc(
  path_10x = "/gpfs0/home/gdrobertslab/lab/Counts/S0052/outs/filtered_feature_bc_matrix/",
  species_pattern = "^hg19-")

t143b_tib_raw <- subset(t143b_tib_raw,
  subset = nFeature_RNA > 1000 &
  nCount_RNA < 30000 &
  percent.mt < 22)

t143b_tib_raw$src <- "Tibia"
t143b_tib_raw$model <- "143B"

# Lung
t143b_lung_raw <- tenx_load_qc(
  path_10x = "/gpfs0/home/gdrobertslab/lab/Counts/S0019-143B-lung/outs/filtered_feature_bc_matrix/",
  species_pattern = "^hg19-")

t143b_lung_raw <- subset(t143b_lung_raw,
  subset = nFeature_RNA > 1000 &
  nCount_RNA < 30000 &
  percent.mt < 22)

t143b_lung_raw$src <- "Lung"
t143b_lung_raw$model <- "143B"
```

## Load NCH-OS2

```{r load_nchos2}
# NCHOS2
## Flank
os2_cx_raw <- tenx_load_qc(
  path_10x = "/gpfs0/home/gdrobertslab/lab/Counts/S0076/outs/filtered_feature_bc_matrix/",
  species_pattern = "^hg19-")

os2_cx_raw <- subset(os2_cx_raw,
  subset = nCount_RNA < 36000 &
  percent.mt < 50)

os2_cx_raw$src <- "Flank"
os2_cx_raw$model <- "NCH-OS-2"

## Tibia
os2_tib_raw<- tenx_load_qc(
  path_10x = "/gpfs0/home/gdrobertslab/lab/Counts/S0042/outs/filtered_feature_bc_matrix/",
  species_pattern = "^hg19-")

os2_tib_raw <- subset(os2_tib_raw,
  subset = nFeature_RNA > 2500 &
  nCount_RNA < 40000 &
  percent.mt < 20)

os2_tib_raw$src <- "Tibia"
os2_tib_raw$model <- "NCH-OS-2"

## Lung
os2_lung_raw<- tenx_load_qc(
  path_10x = "/gpfs0/home/gdrobertslab/lab/Counts/S0041/outs/filtered_feature_bc_matrix/",
  species_pattern = "^hg19-")

os2_lung_raw <- subset(os2_lung_raw,
  subset = nFeature_RNA > 2500 &
  nCount_RNA < 60000 &
  percent.mt < 30)

os2_lung_raw$src <- "Lung"
os2_lung_raw$model <- "NCH-OS-2"
```

## Load NCH-OS7

```{r load_nchos7}
# NCHOS7
## Flank
os7_cx_raw <- tenx_load_qc(
  path_10x = "/gpfs0/home/gdrobertslab/lab/Counts/S0043/outs/filtered_feature_bc_matrix/",
  species_pattern = "^hg19-")

os7_cx_raw <- subset(os7_cx_raw,
  subset = nCount_RNA < 50000 &
  percent.mt < 25)

os7_cx_raw$src <- "Flank"
os7_cx_raw$model <- "NCH-OS-7"

## Tibia
os7_tib_raw <- tenx_load_qc(
  path_10x = "/gpfs0/home/gdrobertslab/lab/Counts/S0034/outs/filtered_feature_bc_matrix/",
  species_pattern = "^hg19-")

os7_tib_raw <- subset(os7_tib_raw,
  subset = nFeature_RNA > 2000 &
  nCount_RNA < 35000 &
  percent.mt < 14)

os7_tib_raw$src <- "Tibia"
os7_tib_raw$model <- "NCH-OS-7"

## Lung
os7_lung_raw <- tenx_load_qc(
  path_10x = "/gpfs0/home/gdrobertslab/lab/Counts/S0055/outs/filtered_feature_bc_matrix/",
  species_pattern = "^hg19-")

os7_lung_raw <- subset(os7_lung_raw,
  subset = nCount_RNA < 42000 &
  percent.mt < 20)

os7_lung_raw$src <- "Lung"
os7_lung_raw$model <- "NCH-OS-7"
```

### Process NCHS-OS7 flank for Figure 1

```{r Fig1a3_process, dependson='load_nchos7', fig.height=8, fig.width=8}
os7_cx <- NormalizeData(os7_cx_raw) %>%
  FindVariableFeatures(selection.method = "vst") %>%
  ScaleData() %>%
  RunPCA(pc.genes = os7_cx@var.genes, npcs = 20) %>%
  RunUMAP(reduction = "pca", dims = 1:20) %>%
  FindNeighbors(reduction = "pca", dims = 1:20) %>%
  FindClusters(resolution = 0.3)

DimPlot(os7_cx,
    reduction = "umap",
    pt.size = 1,
    label = T) +
  coord_fixed() +
  scale_color_npg(alpha = 0.7)

# Attempt to regress out the effects of cell cycle on these tumor cells
s.genes <- cc.genes$s.genes
g2m.genes <- cc.genes$g2m.genes

os7_cx <- CellCycleScoring(object = os7_cx,
  s.features = s.genes,
  g2m.features = g2m.genes,
  set.ident = TRUE)

os7_cx <- ScaleData(object = os7_cx,
  vars.to.regress = c("S.Score", "G2M.Score"),
  features = VariableFeatures(os7_cx),
  block.size = 10000)

os7_cx <- RunPCA(os7_cx, pc.genes = os7_cx@var.genes, npcs = 20) %>%
  RunUMAP(reduction = "pca", dims = 1:20) %>%
  FindNeighbors(reduction = "pca", dims = 1:20) %>%
  FindClusters(resolution = 0.3)

#Find optimal clustering resolution
os7_cx_nc <- nRes(os7_cx, 
  res = seq(from = 0.1, to = 0.15, by = 0.01))
plot <- pSil(os7_cx_nc, 0.15)
plot

os7_cx <- FindClusters(os7_cx, resolution = 0.15)

save(os7_cx, file = "Data/os7_cx_CCR.RData")
```

## Save Raw Objects

```{r save_raw, dependson=c('load_ob','load_os17','load_t143b','load_nchos2','load_nchos7')}
# Save raw objects

# Make lists for easy loop saving
sample_list <- c(
  osteoblasts,
  os17_cx_raw,
  os17_tib_raw,
  os17_lung_raw,
  t143b_cx_raw,
  t143b_tib_raw,
  t143b_lung_raw,
  os2_cx_raw,
  os2_tib_raw,
  os2_lung_raw,
  os7_cx_raw,
  os7_tib_raw,
  os7_lung_raw)

sample_names <- c(
  "osteoblasts",
  "os17_cx_raw",
  "os17_tib_raw",
  "os17_lung_raw",
  "t143b_cx_raw",
  "t143b_tib_raw",
  "t143b_lung_raw",
  "os2_cx_raw",
  "os2_tib_raw",
  "os2_lung_raw",
  "os7_cx_raw",
  "os7_tib_raw",
  "os7_lung_raw")

# Correlate sample names and sample labels
names(sample_list) <- sample_names

# Save each raw object
for (item in sample_names) {
  # Save each sample as an individual Seurat object with proper name
  assign(item, sample_list[[item]])
  save(list = item, file = paste("Data/",
    item,
    ".RData",
    sep = ""))
}
```

# Merge Objects

Merge objects, process, and cell cycle regress. 

```{r mergeallobj, dependson=c('load_ob','load_os17','load_t143b','load_nchos2','load_nchos7'), fig.height=8, fig.width=8}
# Merge into a single Seurat object
OS <- merge(osteoblasts,
  y = c(os17_cx_raw,
    os17_tib_raw,
    os17_lung_raw,
    t143b_cx_raw,
    t143b_tib_raw,
    t143b_lung_raw,
    os2_cx_raw,
    os2_tib_raw,
    os2_lung_raw,
    os7_cx_raw,
    os7_tib_raw,
    os7_lung_raw),
  add.cell.ids = c(
    "Osteoblasts",
    "OS-17_Culture",
    "OS-17_Tibia",
    "OS-17_Lung",
    "143B_Culture",
    "143B_Tibia",
    "143B_Lung",
    "NCH-OS-2_Flank",
    "NCH-OS-2_Tibia",
    "NCH-OS-2_Lung",
    "NCH-OS-7_Flank",
    "NCH-OS-7_Tibia",
    "NCH-OS-7_Lung"),
  project = "Heterogeneity")

# Process and cluster
OS <- NormalizeData(OS) %>%
  FindVariableFeatures(selection.method = "vst") %>%
  ScaleData() %>%
  RunPCA(pc.genes = os.17@var.genes, npcs = 20) %>%
  RunUMAP(reduction = "pca", dims = 1:20) %>%
  FindNeighbors(reduction = "pca", dims = 1:20) %>%
  FindClusters(resolution = 0.3)

# rm(osteoblasts, os17_cx_raw, os17_tib_raw, os17_lung_raw, t143b_cx_raw, t143b_tib_raw,
# t143b_lung_raw, os2_cx_raw, os2_tib_raw, os2_lung_raw, os7_cx_raw, os7_tib_raw,
# os7_lung_raw)

# CCR Attempt to regress out the effects of cell cycle on these tumor cells
s.genes <- cc.genes$s.genes
g2m.genes <- cc.genes$g2m.genes

OS <- CellCycleScoring(object = OS,
  s.features = s.genes,
  g2m.features = g2m.genes,
  set.ident = TRUE)

OS <- ScaleData(object = OS,
  vars.to.regress = c("S.Score", "G2M.Score"),
  features = VariableFeatures(OS),
  block.size = 10000)

OS <- RunPCA(OS, pc.genes = OS@var.genes, npcs = 20) %>%
  RunUMAP(reduction = "pca", dims = 1:20) %>%
  FindNeighbors(reduction = "pca", dims = 1:20) %>%
  FindClusters(resolution = 0.3)

save(OS, file = "Data/OS_merged_postCCR.RData")

OS$model <- as.factor(OS$model)
OS$model <- factor(as.factor(OS$model),
  levels = c(
    "OS-17",
    "143B",
    "NCH-OS-2",
    "NCH-OS-7",
    "Osteoblasts"))

# Plot the data
DimPlot(OS,
    reduction = "umap",
    group.by = "model",
    pt.size = 1,
    label = T) +
  coord_fixed() +
  ggtitle("OS by Model") +
    scale_color_npg(alpha = 1)

DimPlot(OS,
    reduction = "umap",
    group.by = "src",
    pt.size = 1,
    label = T) +
  coord_fixed() +
  ggtitle("OS by Source")

DimPlot(OS,
    reduction = "umap",
    pt.size = 1,
    label = T) +
  coord_fixed() +
  ggtitle("OS by Clusters") +
    scale_color_npg(alpha = 0.7)
```

# List Objects and Process

List objects, process, and cell cycle regress. Used in Figure 3B - ITH code.

```{r listallobj1500, dependson=c('load_ob','load_os17','load_t143b','load_nchos2','load_nchos7')}
# Create List to compute ITH scores (subset to 1500 cells for each)
OS.list_1500 <- list(
  osteoblasts = osteoblasts,
  OS17.cx = os17_cx_raw,
  OS17.Tibia = os17_tib_raw, 
  OS17.Lung = os17_lung_raw, 
  t143B.cx = t143b_cx_raw,
  t143B.Tibia = t143b_tib_raw, 
  t143B.Lung = t143b_lung_raw,
  OS2.Flank = os2_cx_raw,
  OS2.Tibia = os2_tib_raw, 
  OS2.Lung = os2_lung_raw, 
  OS7.Flank= os7_cx_raw,
  OS7.Tibia = os7_tib_raw, 
  OS7.Lung = os7_lung_raw) 

for (i in seq_along(OS.list_1500)) {
  OS.list_1500[[i]] <- NormalizeData(OS.list_1500[[i]]) %>%
    FindVariableFeatures(selection.method = "vst") %>%
    ScaleData() %>%
    RunPCA(pc.genes = OS.list_1500[[i]]@var.genes, npcs = 20) %>%
    RunUMAP(reduction = "pca", dims = 1:20) %>%
    FindNeighbors(reduction = "pca", dims = 1:20) %>%
    FindClusters(resolution = 0.3) %>%
    subset(cells = sample(Cells(OS.list_1500[[i]]), 1500))
}

# CCR
# Attempt to regress out the effects of cell cycle on these tumor cells
s.genes <- cc.genes$s.genes
g2m.genes <- cc.genes$g2m.genes

for (i in seq_along(OS.list_1500)) {
  OS.list_1500[[i]] <- CellCycleScoring(object = OS.list_1500[[i]],
    s.features = s.genes,
    g2m.features = g2m.genes,
    set.ident = TRUE)
  
  OS.list_1500[[i]] <- ScaleData(object = OS.list_1500[[i]],
    vars.to.regress = c("S.Score", "G2M.Score"),
    features = VariableFeatures(OS.list_1500[[i]]),
    block.size = 10000) # Change to VariableFeatures()
  
  OS.list_1500[[i]] <- RunPCA(OS.list_1500[[i]],
                         pc.genes = OS.list_1500[[i]]@var.genes, npcs = 20) %>%
    RunUMAP(reduction = "pca", dims = 1:20) %>%
    FindNeighbors(reduction = "pca", dims = 1:20) %>%
    FindClusters(resolution = 0.3)
}

save(OS.list_1500,
     file = "Data/OSlist_1500_CCR.RData")
```

List objects, process, and cell cycle regress. Used in Figure 3 C-D code.

```{r listallobj, dependson=c('load_ob','load_os17','load_t143b','load_nchos2','load_nchos7')}
# Create list for usage in Figure 3
tibia_list <- list(OS17.Tibia = os17_tib_raw,
  t143B.Tibia = t143b_tib_raw,
  OS2.Tibia = os2_tib_raw,
  OS7.Tibia = os7_tib_raw)

lung_list <- list(OS17.Lung = os17_lung_raw,
  t143B.Lung = t143b_lung_raw,
  OS2.Lung = os2_lung_raw,
  OS7.Lung = os7_lung_raw)

OS_list <- list()

for (i in seq_along(tibia_list)) {
  message(i, head(paste(tibia_list[[i]]$model, "_TL", sep = ""), 1))
  # Subset number of cells in each Seurat object to the least in the group
  a <- table(tibia_list[[i]]$orig.ident)
  b <- table(lung_list[[i]]$orig.ident)
  list <- c(a, b)
  min <- min(list)

  tibia_list[[i]] <- subset(
    tibia_list[[i]],
    cells = sample(Cells(tibia_list[[i]]),
    min))

  lung_list[[i]] <- subset(
    lung_list[[i]],
    cells = sample(Cells(lung_list[[i]]),
    min))

  title <- head(paste(tibia_list[[i]]$model, "_TL", sep = ""), 1)
  title <- gsub("-", "", title)

  OS_list[[title]] <- merge(
    tibia_list[[i]],
    y = c(lung_list[[i]]),
    add.cell.ids = c(
      head(paste(tibia_list[[i]]$model, "_", tibia_list[[i]]$src, sep = ""), 1),
      head(paste(lung_list[[i]]$model, "_", lung_list[[i]]$src, sep = ""), 1)),
      project = "Heterogeneity")
}
```

```{r listallobjccr, dependson=c('load_ob','load_os17','load_t143b','load_nchos2','load_nchos7','listallobj')}
# CCR
# Attempt to regress out the effects of cell cycle on these tumor cells
s.genes <- cc.genes$s.genes
g2m.genes <- cc.genes$g2m.genes

for (i in seq_along(OS_list)) {
  OS_list[[i]] <- NormalizeData(OS_list[[i]]) %>%
    FindVariableFeatures(selection.method = "vst") %>%
    ScaleData() %>%
    RunPCA(pc.genes = OS_list[[i]]@var.genes, npcs = 20) %>%
    RunUMAP(reduction = "pca", dims = 1:20) %>%
    FindNeighbors(reduction = "pca", dims = 1:20) %>%
    FindClusters(resolution = 0.3)

  OS_list[[i]] <- CellCycleScoring(object = OS_list[[i]],
    s.features = s.genes,
    g2m.features = g2m.genes,
    set.ident = TRUE)

  OS_list[[i]] <- ScaleData(object = OS_list[[i]],
    vars.to.regress = c("S.Score", "G2M.Score"),
    features = VariableFeatures(OS_list[[i]]),
    block.size = 10000) # Change to VariableFeatures()

  OS_list[[i]] <- RunPCA(OS_list[[i]],
    pc.genes = OS_list[[i]]@var.genes, npcs = 20) %>%
    RunUMAP(reduction = "pca", dims = 1:20) %>%
    FindNeighbors(reduction = "pca", dims = 1:20) %>%
    FindClusters(resolution = 0.3)
}

save(OS_list, file = "Data/OS_list_CCR.RData")
```

```{r versions}
sessionInfo()
```